{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add Clinvar pathogenic scores to variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8_/3b98_7m10n32rvh6mgvypgj80000gn/T/ipykernel_16308/3302932304.py:2: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"/Users/uriel/Downloads/work_temp/variant_summary.txt\", sep = '\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         #AlleleID                       Type  \\\n",
      "0            15041                      Indel   \n",
      "1            15041                      Indel   \n",
      "2            15042                   Deletion   \n",
      "3            15042                   Deletion   \n",
      "4            15043  single nucleotide variant   \n",
      "...            ...                        ...   \n",
      "4800558    2842239                      Indel   \n",
      "4800559    2842240  single nucleotide variant   \n",
      "4800560    2842240  single nucleotide variant   \n",
      "4800561    2842773                   Deletion   \n",
      "4800562    2842773                   Deletion   \n",
      "\n",
      "                                                      Name  GeneID GeneSymbol  \\\n",
      "0        NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
      "1        NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
      "2           NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
      "3           NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
      "4             NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
      "...                                                    ...     ...        ...   \n",
      "4800558       NM_000030.3(AGXT):c.830delinsAA (p.Ala277fs)     189       AGXT   \n",
      "4800559            NM_000030.3(AGXT):c.139G>C (p.Gly47Arg)     189       AGXT   \n",
      "4800560            NM_000030.3(AGXT):c.139G>C (p.Gly47Arg)     189       AGXT   \n",
      "4800561  NM_020638.3(FGF23):c.515_529del (p.Pro172_Arg1...    8074      FGF23   \n",
      "4800562  NM_020638.3(FGF23):c.515_529del (p.Pro172_Arg1...    8074      FGF23   \n",
      "\n",
      "            HGNC_ID    ClinicalSignificance  ClinSigSimple LastEvaluated  \\\n",
      "0        HGNC:22197              Pathogenic              1             -   \n",
      "1        HGNC:22197              Pathogenic              1             -   \n",
      "2        HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
      "3        HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
      "4        HGNC:28986  Uncertain significance              0  Jun 29, 2015   \n",
      "...             ...                     ...            ...           ...   \n",
      "4800558    HGNC:341              Pathogenic              1  Oct 27, 2023   \n",
      "4800559    HGNC:341       Likely pathogenic              1  Oct 27, 2023   \n",
      "4800560    HGNC:341       Likely pathogenic              1  Oct 27, 2023   \n",
      "4800561   HGNC:3680       Likely pathogenic              1             -   \n",
      "4800562   HGNC:3680       Likely pathogenic              1             -   \n",
      "\n",
      "         RS# (dbSNP)  ...                         ReviewStatus  \\\n",
      "0          397704705  ...  criteria provided, single submitter   \n",
      "1          397704705  ...  criteria provided, single submitter   \n",
      "2          397704709  ...       no assertion criteria provided   \n",
      "3          397704709  ...       no assertion criteria provided   \n",
      "4          150829393  ...       no assertion criteria provided   \n",
      "...              ...  ...                                  ...   \n",
      "4800558           -1  ...  criteria provided, single submitter   \n",
      "4800559           -1  ...  criteria provided, single submitter   \n",
      "4800560           -1  ...  criteria provided, single submitter   \n",
      "4800561           -1  ...  criteria provided, single submitter   \n",
      "4800562           -1  ...  criteria provided, single submitter   \n",
      "\n",
      "        NumberSubmitters Guidelines TestedInGTR  \\\n",
      "0                      2          -           N   \n",
      "1                      2          -           N   \n",
      "2                      1          -           N   \n",
      "3                      1          -           N   \n",
      "4                      1          -           N   \n",
      "...                  ...        ...         ...   \n",
      "4800558                1          -           N   \n",
      "4800559                1          -           N   \n",
      "4800560                1          -           N   \n",
      "4800561                1          -           N   \n",
      "4800562                1          -           N   \n",
      "\n",
      "                                                  OtherIDs  \\\n",
      "0                        ClinGen:CA215070,OMIM:613653.0001   \n",
      "1                        ClinGen:CA215070,OMIM:613653.0001   \n",
      "2                        ClinGen:CA215072,OMIM:613653.0002   \n",
      "3                        ClinGen:CA215072,OMIM:613653.0002   \n",
      "4        ClinGen:CA210674,UniProtKB:Q92610#VAR_064583,O...   \n",
      "...                                                    ...   \n",
      "4800558                                                  -   \n",
      "4800559                                                  -   \n",
      "4800560                                                  -   \n",
      "4800561                                                  -   \n",
      "4800562                                                  -   \n",
      "\n",
      "        SubmitterCategories VariationID PositionVCF ReferenceAlleleVCF  \\\n",
      "0                         3           2     4820844               GGAT   \n",
      "1                         3           2     4781213               GGAT   \n",
      "2                         1           3     4827360    GCTGCTGGACCTGCC   \n",
      "3                         1           3     4787729    GCTGCTGGACCTGCC   \n",
      "4                         1           4    85342440                  G   \n",
      "...                     ...         ...         ...                ...   \n",
      "4800558                   2     2681192   240875988                  C   \n",
      "4800559                   2     2681194   241808421                  G   \n",
      "4800560                   2     2681194   240869004                  G   \n",
      "4800561                   2     2681728     4479735   TGCCGCCGTGGTATGG   \n",
      "4800562                   2     2681728     4370569   TGCCGCCGTGGTATGG   \n",
      "\n",
      "             AlternateAlleleVCF  \n",
      "0        TGCTGTAAACTGTAACTGTAAA  \n",
      "1        TGCTGTAAACTGTAACTGTAAA  \n",
      "2                             G  \n",
      "3                             G  \n",
      "4                             A  \n",
      "...                         ...  \n",
      "4800558                      AA  \n",
      "4800559                       C  \n",
      "4800560                       C  \n",
      "4800561                       T  \n",
      "4800562                       T  \n",
      "\n",
      "[4800563 rows x 34 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"/Users/uriel/Downloads/work_temp/variant_summary.txt\", sep = '\\t')\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>4820844</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>4781213</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>4827361</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>4787730</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4800558</th>\n",
       "      <td>2</td>\n",
       "      <td>240875988</td>\n",
       "      <td>C</td>\n",
       "      <td>AA</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4800559</th>\n",
       "      <td>2</td>\n",
       "      <td>241808421</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4800560</th>\n",
       "      <td>2</td>\n",
       "      <td>240869004</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4800561</th>\n",
       "      <td>12</td>\n",
       "      <td>4479736</td>\n",
       "      <td>TGCCGCCGTGGTATGG</td>\n",
       "      <td>T</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4800562</th>\n",
       "      <td>12</td>\n",
       "      <td>4370570</td>\n",
       "      <td>TGCCGCCGTGGTATGG</td>\n",
       "      <td>T</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4800563 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chrom        Pos               Ref                     Alt  \\\n",
       "0           7    4820844              GGAT  TGCTGTAAACTGTAACTGTAAA   \n",
       "1           7    4781213              GGAT  TGCTGTAAACTGTAACTGTAAA   \n",
       "2           7    4827361   GCTGCTGGACCTGCC                       G   \n",
       "3           7    4787730   GCTGCTGGACCTGCC                       G   \n",
       "4          15   85342440                 G                       A   \n",
       "...       ...        ...               ...                     ...   \n",
       "4800558     2  240875988                 C                      AA   \n",
       "4800559     2  241808421                 G                       C   \n",
       "4800560     2  240869004                 G                       C   \n",
       "4800561    12    4479736  TGCCGCCGTGGTATGG                       T   \n",
       "4800562    12    4370570  TGCCGCCGTGGTATGG                       T   \n",
       "\n",
       "           ClinicalSignificance  \n",
       "0                    Pathogenic  \n",
       "1                    Pathogenic  \n",
       "2                    Pathogenic  \n",
       "3                    Pathogenic  \n",
       "4        Uncertain significance  \n",
       "...                         ...  \n",
       "4800558              Pathogenic  \n",
       "4800559       Likely pathogenic  \n",
       "4800560       Likely pathogenic  \n",
       "4800561       Likely pathogenic  \n",
       "4800562       Likely pathogenic  \n",
       "\n",
       "[4800563 rows x 5 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.rename(columns={\"Chromosome\": \"Chrom\", \"Start\": \"Pos\", \"ReferenceAlleleVCF\": \"Ref\", \"AlternateAlleleVCF\": \"Alt\"})\n",
    "df = df[[\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"ClinicalSignificance\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chrom                   object\n",
       "Pos                      int64\n",
       "Ref                     object\n",
       "Alt                     object\n",
       "ClinicalSignificance    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ACTN2_var = pd.read_csv(\"/Users/uriel/Downloads/work_temp/variant_files/ACTN2.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged = pd.merge(ACTN2_var, df, how = 'inner', on = ['Pos', 'Ref', 'Alt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Chrom_x</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Name</th>\n",
       "      <th>vep_consequence</th>\n",
       "      <th>GERP++_RS</th>\n",
       "      <th>trv</th>\n",
       "      <th>CADD_raw</th>\n",
       "      <th>phyloP100way_vertebrate</th>\n",
       "      <th>Chrom_y</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236717855.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1-236717855-C-A</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236717861.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1-236717861-T-C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.398764</td>\n",
       "      <td>8.011</td>\n",
       "      <td>1</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236717864.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1-236717864-A-G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.935053</td>\n",
       "      <td>9.318</td>\n",
       "      <td>1</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236717897.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1-236717897-G-A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.500607</td>\n",
       "      <td>8.154</td>\n",
       "      <td>1</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236717906.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1-236717906-A-G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.903119</td>\n",
       "      <td>9.325</td>\n",
       "      <td>1</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>760</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762593.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1-236762593-G-A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.43</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.291995</td>\n",
       "      <td>9.812</td>\n",
       "      <td>1</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>763</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762596.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1-236762596-C-T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.43</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.152871</td>\n",
       "      <td>3.301</td>\n",
       "      <td>1</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>767</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762602.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1-236762602-G-A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.990784</td>\n",
       "      <td>9.812</td>\n",
       "      <td>1</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>771</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762611.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1-236762611-G-A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.454363</td>\n",
       "      <td>9.812</td>\n",
       "      <td>1</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>772</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762612.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1-236762612-A-G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5.42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.017029</td>\n",
       "      <td>9.149</td>\n",
       "      <td>1</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>280 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0 Chrom_x          Pos Ref Alt             Name  \\\n",
       "0             0    chr1  236717855.0   C   A  1-236717855-C-A   \n",
       "1             2    chr1  236717861.0   T   C  1-236717861-T-C   \n",
       "2             4    chr1  236717864.0   A   G  1-236717864-A-G   \n",
       "3            12    chr1  236717897.0   G   A  1-236717897-G-A   \n",
       "4            15    chr1  236717906.0   A   G  1-236717906-A-G   \n",
       "..          ...     ...          ...  ..  ..              ...   \n",
       "275         760    chr1  236762593.0   G   A  1-236762593-G-A   \n",
       "276         763    chr1  236762596.0   C   T  1-236762596-C-T   \n",
       "277         767    chr1  236762602.0   G   A  1-236762602-G-A   \n",
       "278         771    chr1  236762611.0   G   A  1-236762611-G-A   \n",
       "279         772    chr1  236762612.0   A   G  1-236762612-A-G   \n",
       "\n",
       "      vep_consequence  GERP++_RS  trv  CADD_raw  phyloP100way_vertebrate  \\\n",
       "0         deleterious        NaN  1.0       NaN                      NaN   \n",
       "1    missense_variant       5.61  0.0  4.398764                    8.011   \n",
       "2    missense_variant       5.61  0.0  3.935053                    9.318   \n",
       "3    missense_variant       5.61  0.0  2.500607                    8.154   \n",
       "4    missense_variant       5.61  0.0  2.903119                    9.325   \n",
       "..                ...        ...  ...       ...                      ...   \n",
       "275  missense_variant       5.43  0.0  2.291995                    9.812   \n",
       "276  missense_variant       5.43  0.0  3.152871                    3.301   \n",
       "277  missense_variant       5.42  0.0  3.990784                    9.812   \n",
       "278  missense_variant       5.42  0.0  3.454363                    9.812   \n",
       "279  missense_variant       5.42  0.0  4.017029                    9.149   \n",
       "\n",
       "    Chrom_y                          ClinicalSignificance  \n",
       "0         1                        Uncertain significance  \n",
       "1         1                        Uncertain significance  \n",
       "2         1                        Uncertain significance  \n",
       "3         1  Conflicting interpretations of pathogenicity  \n",
       "4         1                        Uncertain significance  \n",
       "..      ...                                           ...  \n",
       "275       1  Conflicting interpretations of pathogenicity  \n",
       "276       1                        Uncertain significance  \n",
       "277       1  Conflicting interpretations of pathogenicity  \n",
       "278       1  Conflicting interpretations of pathogenicity  \n",
       "279       1                        Uncertain significance  \n",
       "\n",
       "[280 rows x 13 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
