{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hypertrophic Cardiomyopathy Genes Cross-Validation\n",
    "##### Selin Kubali\n",
    "##### 12/13/2023\n",
    "## Goal\n",
    "Find out whether we can distinguish the HCM risk of bottom 25% and top 25% of missense and deleterious variant carriers in key hypertrophic cardiomyopathy-related genes.\n",
    "\n",
    "#### How the code functions\n",
    "Use cross-validation to fit a Cox-PH model and predict hazard scores. Then isolate the bottom 25% and top 25% of carriers by hazard score and calculate whether there is a statistically significant difference in HCM between them use the Mann-Whitney U test.\n",
    "\n",
    "Cross-validation is done by splitting on variant data, to ensure there are an equal number of variants in each fold and prevent overfitting on high-frequency variants.\n",
    "\n",
    "#### Inputs\n",
    "Lifelines files - from running generate_extracts_gnomAD.ipynb on UKBiobank in Cassa Lab Shared Project/selected_genes/hcm/notebooks. Stored in Cassa Lab Shared Project/selected_genes/hcm/lifelines_data. \n",
    "Variant data files - from running vep_processing.ipynb on UKBiobank in Cassa Lab Shared Project/selected_genes/hcm/notebooks. Stored in Cassa Lab Shared Project/selected_genes/hcm/parsed_vep_files\n",
    "\n",
    "#### Note\n",
    "Two HCM related genes - DES and PLN - were eliminated for having too few variants to converge.\n",
    "PTPN11, TNNI3, and TTR each have few cases of HCM with missense or deleterious variants, which may harm convergence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from lifelines import CoxPHFitter\n",
    "from sklearn.model_selection import KFold\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import warnings; warnings.simplefilter('ignore')\n",
    "import matplotlib.pyplot as plt \n",
    "import numpy as np "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_val(gene):\n",
    "        cph = CoxPHFitter(penalizer=0.0000001)\n",
    "        # load lifelines file\n",
    "        file_name=gene+'.csv'\n",
    "        lifelines_data = pd.read_csv(\"/Users/uriel/Downloads/work_temp/cross_val_lifelines/\"+file_name, dtype={\n",
    "                'is_family_hist':'boolean',\n",
    "                'is_hcm':'boolean'\n",
    "                })\n",
    "\n",
    "        # load variant data file\n",
    "        file_name=gene+'.csv'\n",
    "        variant_data = pd.read_csv(\"/Users/uriel/Downloads/work_temp/variant_files/\"+file_name)\n",
    "\n",
    "\n",
    "        clinvar_df = pd.read_csv(\"/Users/uriel/Downloads/work_temp/variant_summary.txt\", sep = '\\t')\n",
    "        clinvar_df = clinvar_df.rename(columns={\"Chromosome\": \"Chrom\", \"Start\": \"Pos\", \"ReferenceAlleleVCF\": \"Ref\", \"AlternateAlleleVCF\": \"Alt\"})\n",
    "        clinvar_df = clinvar_df[[\"Name\", \"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"ClinicalSignificance\"]]\n",
    "        clinvar_variant_df = pd.merge(variant_data, clinvar_df, how = 'left', on = ['Pos', 'Ref', 'Alt'])\n",
    "        clinvar_variant_df = clinvar_variant_df.rename(columns={\"Name_x\": \"Name\"})\n",
    "        clinvar_variant_df = clinvar_variant_df[[\"Name\", \"ClinicalSignificance\"]]\n",
    "        lifelines_data_with_clin_sign = pd.merge(clinvar_variant_df, lifelines_data, how=\"right\", on = ['Name'])\n",
    "        lifelines_data_with_clin_sign = lifelines_data_with_clin_sign.drop(['Name', 'Carrier', 'index', 'am_pathogenicity'], axis = 1)\n",
    "\n",
    "\n",
    "\n",
    "        variant_data = variant_data[['Name']]\n",
    "        variant_data['var_index'] = variant_data.index\n",
    "\n",
    "\n",
    "\n",
    "        # set lifelines data index to variant data index\n",
    "        lifelines_data = variant_data.merge(lifelines_data, how=\"outer\")\n",
    "        lifelines_data.set_index(\"var_index\")\n",
    "\n",
    "        # clean lifelines file; set pathogenicity for deleterious variants to 1\n",
    "        lifelines_data.loc[lifelines_data['deleterious'] == 1, 'am_pathogenicity'] = 1\n",
    "        lifelines_data = lifelines_data.drop(['Name','Carrier', 'index', 'am_pathogenicity'], axis = 1)\n",
    "        #lifelines_data = lifelines_data[(lifelines_data['deleterious'] == True) | (lifelines_data['missense_variant'] == True)]\n",
    "        #lifelines_data = lifelines_data.drop(['deleterious','missense_variant', 'synonymous_variant'], axis = 1)\n",
    "\n",
    "        lifelines_data = lifelines_data.dropna()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # cross validation: split up phenotypic data file based on variant file index\n",
    "        kf = KFold(n_splits=5, shuffle=True, random_state=1)\n",
    "        testing_set = []\n",
    "        for train_idx, test_idx in kf.split(variant_data):\n",
    "                train = lifelines_data[lifelines_data['var_index'].isin(train_idx)]\n",
    "                test = lifelines_data[lifelines_data['var_index'].isin(test_idx)]\n",
    "\n",
    "                train = train.drop(['var_index'], axis=1)\n",
    "                test = test.drop(['var_index'], axis=1)\n",
    "\n",
    "                # fit CPH and add hazard scores\n",
    "\n",
    "                \n",
    "                cph.fit(train, duration_col=\"duration\", event_col=\"is_hcm\", fit_options = {\"step_size\":0.1})\n",
    "                hazard_scores_fold = cph.predict_partial_hazard(test)\n",
    "                test['hazard'] = hazard_scores_fold\n",
    "                testing_set.append(test)\n",
    "\n",
    "\n",
    "\n",
    "        # create new lifelines_data df by joining all testing sets\n",
    "        lifelines_data = pd.concat([df for idx, df in enumerate(testing_set)]) \n",
    "        lifelines_data = pd.merge(lifelines_data, lifelines_data_with_clin_sign, how = 'left')\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "        return lifelines_data\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def find_params(gene):\n",
    "        thresholds_list =  list(range(1, 101))\n",
    "\n",
    "\n",
    "\n",
    "        p_vals = {}\n",
    "        hazard_ratios = {}\n",
    "        f1_scores = {}\n",
    "\n",
    "        \n",
    "        lifelines_data = cross_val(gene)\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # filter for patients with lowest 25% and highest 25% hazard scores\n",
    "\n",
    "        for i in thresholds_list:\n",
    "                percentiles = np.percentile(lifelines_data['hazard'], [i])\n",
    "                bottom = lifelines_data[lifelines_data['hazard'] < percentiles[0]]\n",
    "                top = lifelines_data[lifelines_data['hazard'] >= percentiles[0]]\n",
    "\n",
    "                bottom.loc[:,'is_hcm'] = np.where(((bottom['ClinicalSignificance'] == 'Pathogenic') |(bottom['ClinicalSignificance'] == 'Likely pathogenic') | (bottom['ClinicalSignificance'] == 'Pathogenic/Likely pathogenic')) , 1, 0)\n",
    "                top.loc[:,'is_hcm'] = np.where(((top['ClinicalSignificance'] == 'Pathogenic') |(top['ClinicalSignificance'] == 'Likely pathogenic') | (top['ClinicalSignificance'] == 'Pathogenic/Likely pathogenic')), 1, 0)\n",
    "                dfA = pd.DataFrame({'E': bottom['is_hcm'], 'T': bottom['duration'], 'is_highest': 0})\n",
    "                dfB = pd.DataFrame({'E': top['is_hcm'], 'T': top['duration'], 'is_highest': 1})\n",
    "\n",
    "\n",
    "                \"\"\"bottom.loc[:,'is_hcm'] = np.where(bottom['is_hcm'] == True, 1, 0)\n",
    "                top.loc[:,'is_hcm'] = np.where(top['is_hcm'] == True, 1, 0)\n",
    "                dfA = pd.DataFrame({'E': bottom['is_hcm'], 'T': bottom['duration'], 'is_highest': 0})\n",
    "                dfB = pd.DataFrame({'E': top['is_hcm'], 'T': top['duration'], 'is_highest': 1})\"\"\"\n",
    "\n",
    "\n",
    "                df = pd.concat([dfA, dfB])\n",
    "\n",
    "                cph = CoxPHFitter().fit(df, 'T', 'E', fit_options = {\"step_size\":0.1})\n",
    "                hazard_ratios.update({i:cph.hazard_ratios_.at['is_highest']})\n",
    "\n",
    "                p_vals.update({i:cph.summary['p'].at['is_highest']})\n",
    "\n",
    "                TP = (((top['ClinicalSignificance'] == 'Pathogenic') | (top['ClinicalSignificance'] == 'Likely pathogenic') | (top['ClinicalSignificance'] == 'Pathogenic/Likely pathogenic')).sum())\n",
    "                FN = (((bottom['ClinicalSignificance'] == 'Pathogenic') | (bottom['ClinicalSignificance'] == 'Likely pathogenic') | (bottom['ClinicalSignificance'] == 'Pathogenic/Likely pathogenic')).sum())\n",
    "                #FP = (top['synonymous_variant'].sum())\n",
    "                FP = (((top['ClinicalSignificance'] == 'Benign') | (top['ClinicalSignificance'] == 'Likely Benign') | (top['ClinicalSignificance'] == 'Benign/Likely benign')).sum())\n",
    "\n",
    "        \n",
    "                \"\"\"\n",
    "                FN = (bottom['is_hcm'] == 1).sum()\n",
    "                FP = (top['is_hcm'] == 0).sum()\n",
    "                TN = (bottom['is_hcm'] == 0).sum()\n",
    "                \"\"\"\n",
    "\n",
    "\n",
    "\n",
    "                \n",
    "             \n",
    "\n",
    "                #print('Top HCM',top[top[\"is_hcm\"] == 1][\"is_hcm\"].sum())\n",
    "                #print('Bottom HCM', bottom[bottom[\"is_hcm\"] == 1][\"is_hcm\"].sum())\n",
    "\n",
    "\n",
    "                print(TP, FN, FP)\n",
    "\n",
    "                f1_score = TP/(TP+0.5*(FP+FN))\n",
    "                f1_scores.update({i:f1_score})\n",
    "\n",
    "                \n",
    "\n",
    "\n",
    "                # true positive = classified as high risk, ClinVar pathogenic/likely pathogenic\n",
    "                # false positive = classified as low risk, ClinVar pathogenic \n",
    "                # false negative = classified as high risk, synonymous\n",
    "\n",
    "                # find # of low risk syn vs high risk syn, etc.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "        return p_vals, hazard_ratios, f1_scores\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MYH6\n",
      "2 0 931\n",
      "2 0 917\n",
      "2 0 911\n",
      "2 0 909\n",
      "2 0 906\n",
      "2 0 900\n",
      "2 0 895\n",
      "2 0 894\n",
      "2 0 893\n",
      "2 0 888\n",
      "2 0 886\n",
      "2 0 880\n",
      "2 0 874\n",
      "2 0 867\n",
      "2 0 862\n",
      "2 0 858\n",
      "2 0 852\n",
      "1 1 846\n",
      "1 1 841\n",
      "1 1 834\n",
      "1 1 824\n",
      "1 1 810\n",
      "1 1 798\n",
      "1 1 791\n",
      "1 1 789\n",
      "1 1 779\n",
      "1 1 772\n",
      "1 1 769\n",
      "1 1 763\n",
      "1 1 760\n",
      "1 1 748\n",
      "1 1 739\n",
      "1 1 721\n",
      "1 1 713\n",
      "1 1 702\n",
      "1 1 696\n",
      "1 1 688\n",
      "1 1 679\n",
      "1 1 667\n",
      "1 1 650\n",
      "1 1 645\n",
      "1 1 635\n",
      "1 1 622\n",
      "1 1 615\n",
      "1 1 609\n",
      "1 1 601\n",
      "1 1 590\n",
      "1 1 582\n",
      "1 1 571\n",
      "1 1 568\n",
      "1 1 560\n",
      "1 1 550\n",
      "1 1 541\n",
      "1 1 526\n",
      "1 1 515\n",
      "1 1 503\n",
      "1 1 497\n",
      "1 1 486\n",
      "1 1 472\n",
      "1 1 460\n",
      "1 1 448\n",
      "1 1 437\n",
      "1 1 422\n",
      "1 1 413\n",
      "1 1 402\n",
      "1 1 385\n",
      "1 1 378\n",
      "1 1 374\n",
      "1 1 365\n",
      "1 1 359\n",
      "1 1 342\n",
      "1 1 333\n",
      "1 1 323\n",
      "1 1 314\n",
      "1 1 305\n",
      "1 1 291\n",
      "0 2 278\n",
      "0 2 262\n",
      "0 2 251\n",
      "0 2 240\n",
      "0 2 228\n",
      "0 2 220\n",
      "0 2 211\n",
      "0 2 202\n",
      "0 2 192\n",
      "0 2 179\n",
      "0 2 168\n",
      "0 2 154\n",
      "0 2 145\n",
      "0 2 134\n",
      "0 2 121\n",
      "0 2 104\n",
      "0 2 96\n",
      "0 2 85\n",
      "0 2 76\n",
      "0 2 60\n",
      "0 2 50\n",
      "0 2 41\n",
      "0 2 22\n"
     ]
    },
    {
     "ename": "ConvergenceError",
     "evalue": "Convergence halted due to matrix inversion problems. Suspicion is high collinearity. Please see the following tips in the lifelines documentation: https://lifelines.readthedocs.io/en/latest/Examples.html#problems-with-convergence-in-the-cox-proportional-hazard-modelMatrix is singular.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLinAlgError\u001b[0m                               Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1530\u001b[0m, in \u001b[0;36mSemiParametricPHFitter._newton_raphson_for_efron_model\u001b[0;34m(self, X, T, E, weights, entries, initial_point, show_progress, step_size, precision, r_precision, max_steps)\u001b[0m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1530\u001b[0m     inv_h_dot_g_T \u001b[38;5;241m=\u001b[39m \u001b[43mspsolve\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[43mh\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43massume_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpos\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_finite\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m   1531\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mValueError\u001b[39;00m, LinAlgError) \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/scipy/linalg/_basic.py:254\u001b[0m, in \u001b[0;36msolve\u001b[0;34m(a, b, lower, overwrite_a, overwrite_b, check_finite, assume_a, transposed)\u001b[0m\n\u001b[1;32m    251\u001b[0m lu, x, info \u001b[38;5;241m=\u001b[39m posv(a1, b1, lower\u001b[38;5;241m=\u001b[39mlower,\n\u001b[1;32m    252\u001b[0m                    overwrite_a\u001b[38;5;241m=\u001b[39moverwrite_a,\n\u001b[1;32m    253\u001b[0m                    overwrite_b\u001b[38;5;241m=\u001b[39moverwrite_b)\n\u001b[0;32m--> 254\u001b[0m \u001b[43m_solve_check\u001b[49m\u001b[43m(\u001b[49m\u001b[43mn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minfo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    255\u001b[0m rcond, info \u001b[38;5;241m=\u001b[39m pocon(lu, anorm)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/scipy/linalg/_basic.py:41\u001b[0m, in \u001b[0;36m_solve_check\u001b[0;34m(n, info, lamch, rcond)\u001b[0m\n\u001b[1;32m     40\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;241m0\u001b[39m \u001b[38;5;241m<\u001b[39m info:\n\u001b[0;32m---> 41\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m LinAlgError(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMatrix is singular.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     43\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m lamch \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[0;31mLinAlgError\u001b[0m: Matrix is singular.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConvergenceError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[63], line 9\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m genes:\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;28mprint\u001b[39m(gene)\n\u001b[0;32m----> 9\u001b[0m     p_vals, hazard_ratios, f1_scores \u001b[38;5;241m=\u001b[39m \u001b[43mfind_params\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     11\u001b[0m     \u001b[38;5;66;03m# By minimizing p-value\u001b[39;00m\n\u001b[1;32m     12\u001b[0m     associated_p_threshold \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmin\u001b[39m(p_vals, key\u001b[38;5;241m=\u001b[39mp_vals\u001b[38;5;241m.\u001b[39mget)\n",
      "Cell \u001b[0;32mIn[60], line 38\u001b[0m, in \u001b[0;36mfind_params\u001b[0;34m(gene)\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"bottom.loc[:,'is_hcm'] = np.where(bottom['is_hcm'] == True, 1, 0)\u001b[39;00m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;124;03mtop.loc[:,'is_hcm'] = np.where(top['is_hcm'] == True, 1, 0)\u001b[39;00m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;124;03mdfA = pd.DataFrame({'E': bottom['is_hcm'], 'T': bottom['duration'], 'is_highest': 0})\u001b[39;00m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;124;03mdfB = pd.DataFrame({'E': top['is_hcm'], 'T': top['duration'], 'is_highest': 1})\"\"\"\u001b[39;00m\n\u001b[1;32m     36\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([dfA, dfB])\n\u001b[0;32m---> 38\u001b[0m cph \u001b[38;5;241m=\u001b[39m \u001b[43mCoxPHFitter\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mT\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mE\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfit_options\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstep_size\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[38;5;241;43m0.1\u001b[39;49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     39\u001b[0m hazard_ratios\u001b[38;5;241m.\u001b[39mupdate({i:cph\u001b[38;5;241m.\u001b[39mhazard_ratios_\u001b[38;5;241m.\u001b[39mat[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mis_highest\u001b[39m\u001b[38;5;124m'\u001b[39m]})\n\u001b[1;32m     41\u001b[0m p_vals\u001b[38;5;241m.\u001b[39mupdate({i:cph\u001b[38;5;241m.\u001b[39msummary[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mp\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mat[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mis_highest\u001b[39m\u001b[38;5;124m'\u001b[39m]})\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/utils/__init__.py:56\u001b[0m, in \u001b[0;36mCensoringType.right_censoring.<locals>.f\u001b[0;34m(model, *args, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(function)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m(model, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mset_censoring_type(model, \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mRIGHT)\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:290\u001b[0m, in \u001b[0;36mCoxPHFitter.fit\u001b[0;34m(self, df, duration_col, event_col, show_progress, initial_point, strata, weights_col, cluster_col, robust, batch_mode, timeline, formula, entry_col, fit_options)\u001b[0m\n\u001b[1;32m    184\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    185\u001b[0m \u001b[38;5;124;03mFit the Cox proportional hazard model to a right-censored dataset. Alias of `fit_right_censoring`.\u001b[39;00m\n\u001b[1;32m    186\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    287\u001b[0m \n\u001b[1;32m    288\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    289\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata \u001b[38;5;241m=\u001b[39m utils\u001b[38;5;241m.\u001b[39m_to_list_or_singleton(utils\u001b[38;5;241m.\u001b[39mcoalesce(strata, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata))\n\u001b[0;32m--> 290\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_model \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    291\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    292\u001b[0m \u001b[43m    \u001b[49m\u001b[43mduration_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    293\u001b[0m \u001b[43m    \u001b[49m\u001b[43mevent_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mevent_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    294\u001b[0m \u001b[43m    \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mshow_progress\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    295\u001b[0m \u001b[43m    \u001b[49m\u001b[43minitial_point\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_point\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    296\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstrata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstrata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    297\u001b[0m \u001b[43m    \u001b[49m\u001b[43mweights_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mweights_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    298\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcluster_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcluster_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    299\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrobust\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrobust\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    300\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbatch_mode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_mode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    301\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeline\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    302\u001b[0m \u001b[43m    \u001b[49m\u001b[43mformula\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mformula\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    303\u001b[0m \u001b[43m    \u001b[49m\u001b[43mentry_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mentry_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    304\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfit_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfit_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    305\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    306\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:610\u001b[0m, in \u001b[0;36mCoxPHFitter._fit_model\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    608\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_fit_model\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    609\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbaseline_estimation_method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbreslow\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m--> 610\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model_breslow\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    611\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbaseline_estimation_method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mspline\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    612\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fit_model_spline(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:623\u001b[0m, in \u001b[0;36mCoxPHFitter._fit_model_breslow\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    619\u001b[0m model \u001b[38;5;241m=\u001b[39m SemiParametricPHFitter(\n\u001b[1;32m    620\u001b[0m     penalizer\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpenalizer, l1_ratio\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39ml1_ratio, strata\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata, alpha\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39malpha, label\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_label\n\u001b[1;32m    621\u001b[0m )\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m utils\u001b[38;5;241m.\u001b[39mCensoringType\u001b[38;5;241m.\u001b[39mis_right_censoring(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 623\u001b[0m     \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    624\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m model\n\u001b[1;32m    625\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/utils/__init__.py:56\u001b[0m, in \u001b[0;36mCensoringType.right_censoring.<locals>.f\u001b[0;34m(model, *args, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(function)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m(model, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mset_censoring_type(model, \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mRIGHT)\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1252\u001b[0m, in \u001b[0;36mSemiParametricPHFitter.fit\u001b[0;34m(self, df, duration_col, event_col, show_progress, initial_point, strata, weights_col, cluster_col, robust, batch_mode, timeline, formula, entry_col, fit_options)\u001b[0m\n\u001b[1;32m   1247\u001b[0m \u001b[38;5;66;03m# this is surprisingly faster to do...\u001b[39;00m\n\u001b[1;32m   1248\u001b[0m X_norm \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[1;32m   1249\u001b[0m     utils\u001b[38;5;241m.\u001b[39mnormalize(X\u001b[38;5;241m.\u001b[39mvalues, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_norm_mean\u001b[38;5;241m.\u001b[39mvalues, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_norm_std\u001b[38;5;241m.\u001b[39mvalues), index\u001b[38;5;241m=\u001b[39mX\u001b[38;5;241m.\u001b[39mindex, columns\u001b[38;5;241m=\u001b[39mX\u001b[38;5;241m.\u001b[39mcolumns\n\u001b[1;32m   1250\u001b[0m )\n\u001b[0;32m-> 1252\u001b[0m params_, ll_, variance_matrix_, baseline_hazard_, baseline_cumulative_hazard_, model \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1253\u001b[0m \u001b[43m    \u001b[49m\u001b[43mX_norm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1254\u001b[0m \u001b[43m    \u001b[49m\u001b[43mT\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1255\u001b[0m \u001b[43m    \u001b[49m\u001b[43mE\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1256\u001b[0m \u001b[43m    \u001b[49m\u001b[43mweights\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mweights\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1257\u001b[0m \u001b[43m    \u001b[49m\u001b[43mentries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mentries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1258\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfit_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mutils\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcoalesce\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfit_options\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mdict\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1259\u001b[0m \u001b[43m    \u001b[49m\u001b[43minitial_point\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_point\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1260\u001b[0m \u001b[43m    \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mshow_progress\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1261\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1263\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlog_likelihood_ \u001b[38;5;241m=\u001b[39m ll_\n\u001b[1;32m   1264\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel \u001b[38;5;241m=\u001b[39m model\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1380\u001b[0m, in \u001b[0;36mSemiParametricPHFitter._fit_model\u001b[0;34m(self, X, T, E, weights, entries, fit_options, initial_point, show_progress)\u001b[0m\n\u001b[1;32m   1369\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_fit_model\u001b[39m(\n\u001b[1;32m   1370\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1371\u001b[0m     X: DataFrame,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1378\u001b[0m     show_progress: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m   1379\u001b[0m ):\n\u001b[0;32m-> 1380\u001b[0m     beta_, ll_, hessian_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_newton_raphson_for_efron_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1381\u001b[0m \u001b[43m        \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1382\u001b[0m \u001b[43m        \u001b[49m\u001b[43mT\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1383\u001b[0m \u001b[43m        \u001b[49m\u001b[43mE\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1384\u001b[0m \u001b[43m        \u001b[49m\u001b[43mweights\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1385\u001b[0m \u001b[43m        \u001b[49m\u001b[43mentries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1386\u001b[0m \u001b[43m        \u001b[49m\u001b[43minitial_point\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_point\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1387\u001b[0m \u001b[43m        \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mshow_progress\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1388\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mfit_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1389\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1391\u001b[0m     \u001b[38;5;66;03m# compute the baseline hazard here.\u001b[39;00m\n\u001b[1;32m   1392\u001b[0m     predicted_partial_hazards_ \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m   1393\u001b[0m         pd\u001b[38;5;241m.\u001b[39mDataFrame(np\u001b[38;5;241m.\u001b[39mexp(dot(X, beta_)), columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mP\u001b[39m\u001b[38;5;124m\"\u001b[39m])\u001b[38;5;241m.\u001b[39massign(T\u001b[38;5;241m=\u001b[39mT\u001b[38;5;241m.\u001b[39mvalues, E\u001b[38;5;241m=\u001b[39mE\u001b[38;5;241m.\u001b[39mvalues, W\u001b[38;5;241m=\u001b[39mweights\u001b[38;5;241m.\u001b[39mvalues)\u001b[38;5;241m.\u001b[39mset_index(X\u001b[38;5;241m.\u001b[39mindex)\n\u001b[1;32m   1394\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1539\u001b[0m, in \u001b[0;36mSemiParametricPHFitter._newton_raphson_for_efron_model\u001b[0;34m(self, X, T, E, weights, entries, initial_point, show_progress, step_size, precision, r_precision, max_steps)\u001b[0m\n\u001b[1;32m   1534\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exceptions\u001b[38;5;241m.\u001b[39mConvergenceError(\n\u001b[1;32m   1535\u001b[0m \u001b[38;5;250m        \u001b[39m\u001b[38;5;124;03m\"\"\"Hessian or gradient contains nan or inf value(s). Convergence halted. {0}\"\"\"\u001b[39;00m\u001b[38;5;241m.\u001b[39mformat(CONVERGENCE_DOCS),\n\u001b[1;32m   1536\u001b[0m         e,\n\u001b[1;32m   1537\u001b[0m     )\n\u001b[1;32m   1538\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e, LinAlgError):\n\u001b[0;32m-> 1539\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exceptions\u001b[38;5;241m.\u001b[39mConvergenceError(\n\u001b[1;32m   1540\u001b[0m \u001b[38;5;250m        \u001b[39m\u001b[38;5;124;03m\"\"\"Convergence halted due to matrix inversion problems. Suspicion is high collinearity. {0}\"\"\"\u001b[39;00m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m   1541\u001b[0m             CONVERGENCE_DOCS\n\u001b[1;32m   1542\u001b[0m         ),\n\u001b[1;32m   1543\u001b[0m         e,\n\u001b[1;32m   1544\u001b[0m     )\n\u001b[1;32m   1545\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1546\u001b[0m     \u001b[38;5;66;03m# something else?\u001b[39;00m\n\u001b[1;32m   1547\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n",
      "\u001b[0;31mConvergenceError\u001b[0m: Convergence halted due to matrix inversion problems. Suspicion is high collinearity. Please see the following tips in the lifelines documentation: https://lifelines.readthedocs.io/en/latest/Examples.html#problems-with-convergence-in-the-cox-proportional-hazard-modelMatrix is singular."
     ]
    }
   ],
   "source": [
    "genes = ['ACTN2', 'ALPK3', 'FLNC','MYBPC3','MYH6', 'MYH7', 'PTPN11', 'TNNI3', 'TTR']\n",
    "lowest_thresholds_by_p_val = {}\n",
    "thresholds_by_f1_score = {}\n",
    "threshold_by_odds_ratio = {}\n",
    "\n",
    "\n",
    "for gene in genes:\n",
    "    print(gene)\n",
    "    p_vals, hazard_ratios, f1_scores = find_params(gene)\n",
    "    \n",
    "    # By minimizing p-value\n",
    "    associated_p_threshold = min(p_vals, key=p_vals.get)\n",
    "    lowest_thresholds_by_p_val.update({gene:associated_p_threshold})\n",
    "    associated_f1_threshold = max(f1_scores, key=f1_scores.get)\n",
    "    thresholds_by_f1_score.update({gene:associated_f1_threshold})\n",
    "\n",
    "    associated_odds_threshold = max(hazard_ratios, key=hazard_ratios.get)\n",
    "    threshold_by_odds_ratio.update({gene:associated_odds_threshold})\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rerun CoxPH cross-validation with chosen threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_threshold_vals(dict):\n",
    "\n",
    "        p_vals = {}\n",
    "        hazard_ratios = {}\n",
    "        f1_scores = {}\n",
    "\n",
    "\n",
    "        # filter for patients with lowest 25% and highest 25% hazard scores\n",
    "\n",
    "        for gene in dict:\n",
    "                threshold = dict[gene]\n",
    "\n",
    "\n",
    "                lifelines_data = cross_val(gene)\n",
    "\n",
    "\n",
    "\n",
    "                percentiles = np.percentile(lifelines_data['hazard'], [threshold])\n",
    "                bottom = lifelines_data[lifelines_data['hazard'] < percentiles[0]]\n",
    "                top = lifelines_data[lifelines_data['hazard'] >= percentiles[0]]\n",
    "                bottom.loc[:,'is_hcm'] = np.where(bottom['is_hcm'] == True, 1, 0)\n",
    "                top.loc[:,'is_hcm'] = np.where(top['is_hcm'] == True, 1, 0)\n",
    "\n",
    "                print(gene)\n",
    "\n",
    "\n",
    "\n",
    "                dfA = pd.DataFrame({'E': bottom['is_hcm'], 'T': bottom['duration'], 'is_highest': 0})\n",
    "                dfB = pd.DataFrame({'E': top['is_hcm'], 'T': top['duration'], 'is_highest': 1})\n",
    "                df = pd.concat([dfA, dfB])\n",
    "\n",
    "                cph = CoxPHFitter().fit(df, 'T', 'E', fit_options = {\"step_size\":0.1})\n",
    "                hazard_ratios.update({gene:cph.hazard_ratios_.at['is_highest']})\n",
    "\n",
    "                p_vals.update({gene:cph.summary['p'].at['is_highest']})\n",
    "                p_adjusted = multipletests(list(p_vals.values()), alpha=0.05, method='bonferroni')\n",
    "                updated_p_dict = {key: new_p_val for key, new_p_val in zip(p_vals.keys(), p_adjusted[1])}\n",
    "\n",
    "                \"\"\"TP = (((top['ClinicalSignificance'] == 'Pathogenic') | (top['ClinicalSignificance'] == 'Likely pathogenic') | (top['ClinicalSignificance'] == 'Pathogenic/Likely pathogenic')).sum())\n",
    "                FN = (((bottom['ClinicalSignificance'] == 'Pathogenic') | (bottom['ClinicalSignificance'] == 'Likely pathogenic') | (bottom['ClinicalSignificance'] == 'Pathogenic/Likely pathogenic')).sum())\n",
    "                #FP = (top['synonymous_variant'].sum())\n",
    "                FP = (((top['ClinicalSignificance'] == 'Benign') | (top['ClinicalSignificance'] == 'Likely benign') | (top['ClinicalSignificance'] == 'Benign/Likely benign')).sum())\"\"\"\n",
    "\n",
    "                        \n",
    "                TP = (top['is_hcm'] == 1).sum()\n",
    "                FN = (bottom['is_hcm'] == 1).sum()\n",
    "                FP = (top['is_hcm'] == 0).sum()\n",
    "             \n",
    "\n",
    "\n",
    "                f1_score = TP/(TP+0.5*(FP+FN))\n",
    "\n",
    "\n",
    "                f1_scores.update({gene:f1_score})\n",
    "\n",
    "               \n",
    "                \n",
    "\n",
    "                \n",
    "\n",
    "\n",
    "        return updated_p_dict, hazard_ratios, f1_scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert lowest p-values to dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTN2\n",
      "ALPK3\n",
      "FLNC\n",
      "MYBPC3\n",
      "MYH6\n",
      "MYH7\n",
      "PTPN11\n",
      "TNNI3\n",
      "TTR\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P-value</th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>Thresholds</th>\n",
       "      <th>F1 scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACTN2</th>\n",
       "      <td>1.414956e-01</td>\n",
       "      <td>4.063891</td>\n",
       "      <td>91</td>\n",
       "      <td>0.010471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALPK3</th>\n",
       "      <td>8.145572e-04</td>\n",
       "      <td>3.986677</td>\n",
       "      <td>83</td>\n",
       "      <td>0.010518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNC</th>\n",
       "      <td>9.406915e-07</td>\n",
       "      <td>6.811688</td>\n",
       "      <td>83</td>\n",
       "      <td>0.009797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYBPC3</th>\n",
       "      <td>9.209080e-19</td>\n",
       "      <td>31.108449</td>\n",
       "      <td>98</td>\n",
       "      <td>0.076923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH6</th>\n",
       "      <td>2.808500e-08</td>\n",
       "      <td>22.297404</td>\n",
       "      <td>96</td>\n",
       "      <td>0.021239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH7</th>\n",
       "      <td>4.972407e-12</td>\n",
       "      <td>7.440186</td>\n",
       "      <td>89</td>\n",
       "      <td>0.034700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPN11</th>\n",
       "      <td>1.100783e-01</td>\n",
       "      <td>0.068621</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNNI3</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.334120</td>\n",
       "      <td>15</td>\n",
       "      <td>0.007463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTR</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.256571</td>\n",
       "      <td>53</td>\n",
       "      <td>0.008264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             P-value  Odds ratio  Thresholds  F1 scores\n",
       "ACTN2   1.414956e-01    4.063891          91   0.010471\n",
       "ALPK3   8.145572e-04    3.986677          83   0.010518\n",
       "FLNC    9.406915e-07    6.811688          83   0.009797\n",
       "MYBPC3  9.209080e-19   31.108449          98   0.076923\n",
       "MYH6    2.808500e-08   22.297404          96   0.021239\n",
       "MYH7    4.972407e-12    7.440186          89   0.034700\n",
       "PTPN11  1.100783e-01    0.068621           1   0.004496\n",
       "TNNI3   1.000000e+00    0.334120          15   0.007463\n",
       "TTR     1.000000e+00    2.256571          53   0.008264"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "updated_p_dict, hazard_ratios, f1_scores = find_threshold_vals(lowest_thresholds_by_p_val)\n",
    "p_vals = pd.DataFrame.from_dict(updated_p_dict, orient = 'index')\n",
    "p_vals.columns = [\"P-value\"]\n",
    "hazard_ratios = pd.DataFrame.from_dict(hazard_ratios, orient = 'index')\n",
    "hazard_ratios.columns = [\"Odds ratio\"]\n",
    "thresholds = pd.DataFrame.from_dict(lowest_thresholds_by_p_val, orient = 'index')\n",
    "thresholds.columns = [\"Thresholds\"]\n",
    "f1_scores = pd.DataFrame.from_dict(f1_scores, orient = 'index')\n",
    "f1_scores.columns = [\"F1 scores\"]\n",
    "df = p_vals.join(hazard_ratios).join(thresholds).join(f1_scores)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert highest odds ratios to dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0     0.0            True   False  66.368241  66.368241   \n",
      "1     1.0           False   False  82.535250  82.535250   \n",
      "2     1.0           False   False  53.535934  53.535934   \n",
      "3     0.0           False   False  73.618070  73.618070   \n",
      "4     0.0            True   False  61.535934  61.535934   \n",
      "...   ...             ...     ...        ...        ...   \n",
      "8306  1.0            True   False  82.116359  82.116359   \n",
      "8307  0.0            True   False  75.865845  75.865845   \n",
      "8308  0.0            True   False  62.866530  62.866530   \n",
      "8309  1.0           False   False  77.368925  77.368925   \n",
      "8310  1.0           False   False  71.950719  71.950719   \n",
      "\n",
      "      principal_component_1  principal_component_4  prs_score  GERP++_RS  trv  \\\n",
      "0                 -10.16980              -3.208120  -0.002396       6.03  0.0   \n",
      "1                  69.75280              11.018500  -0.000139       6.03  0.0   \n",
      "2                  59.35190               1.404910  -0.007645       6.03  0.0   \n",
      "3                 -15.02350               0.666923  -0.009656       6.03  0.0   \n",
      "4                  80.00940              21.424200  -0.027113       6.03  0.0   \n",
      "...                     ...                    ...        ...        ...  ...   \n",
      "8306               -9.95568               1.462740  -0.022260       6.03  0.0   \n",
      "8307              -11.31300               1.223110  -0.010965       6.03  0.0   \n",
      "8308              -11.30380              -1.633610  -0.010485       5.43  0.0   \n",
      "8309              -13.60130              -1.128440  -0.016980       5.43  0.0   \n",
      "8310              -13.56330               0.015534  -0.009299       5.43  0.0   \n",
      "\n",
      "      CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0     5.864391                   10.003          0.0               0.0   \n",
      "1     5.864391                   10.003          0.0               0.0   \n",
      "2     5.864391                   10.003          0.0               0.0   \n",
      "3     5.864391                   10.003          0.0               0.0   \n",
      "4     5.864391                   10.003          0.0               0.0   \n",
      "...        ...                      ...          ...               ...   \n",
      "8306  5.864391                   10.003          0.0               0.0   \n",
      "8307  5.864391                   10.003          0.0               0.0   \n",
      "8308  3.152871                    3.301          0.0               1.0   \n",
      "8309  3.152871                    3.301          0.0               1.0   \n",
      "8310  3.152871                    3.301          0.0               1.0   \n",
      "\n",
      "      synonymous_variant    hazard  \\\n",
      "0                    1.0  1.842582   \n",
      "1                    1.0  1.784514   \n",
      "2                    1.0  4.738215   \n",
      "3                    1.0  0.195599   \n",
      "4                    1.0  0.918901   \n",
      "...                  ...       ...   \n",
      "8306                 1.0  0.691140   \n",
      "8307                 1.0  0.476379   \n",
      "8308                 0.0  1.510111   \n",
      "8309                 0.0  0.711042   \n",
      "8310                 0.0  1.358027   \n",
      "\n",
      "                              ClinicalSignificance  \n",
      "0                                              NaN  \n",
      "1     Conflicting interpretations of pathogenicity  \n",
      "2     Conflicting interpretations of pathogenicity  \n",
      "3     Conflicting interpretations of pathogenicity  \n",
      "4     Conflicting interpretations of pathogenicity  \n",
      "...                                            ...  \n",
      "8306                                           NaN  \n",
      "8307                                           NaN  \n",
      "8308                        Uncertain significance  \n",
      "8309                        Uncertain significance  \n",
      "8310                        Uncertain significance  \n",
      "\n",
      "[8311 rows x 17 columns]\n",
      "ACTN2\n",
      "       sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0      1.0            True   False  82.283368  82.283368   \n",
      "1      0.0            True   False  71.285421  71.285421   \n",
      "2      1.0           False   False  76.032854  76.032854   \n",
      "3      0.0           False   False  75.619439  75.619439   \n",
      "4      0.0           False   False  67.118412  67.118412   \n",
      "...    ...             ...     ...        ...        ...   \n",
      "14331  1.0           False   False  72.933607  74.951403   \n",
      "14332  1.0           False   False  81.949350  81.949350   \n",
      "14333  0.0            True   False  70.450376  70.450376   \n",
      "14334  0.0            True   False  76.369610  76.369610   \n",
      "14335  1.0            True   False  78.702259  78.702259   \n",
      "\n",
      "       principal_component_1  principal_component_4  prs_score  GERP++_RS  \\\n",
      "0                  -13.91010               3.527340  -0.033126       4.99   \n",
      "1                  -13.71400               5.142310  -0.017023       4.99   \n",
      "2                  -11.31890              -2.350370  -0.011407       4.99   \n",
      "3                   -9.08729               1.119470   0.009379       4.99   \n",
      "4                   -9.12163               0.876826  -0.011631       4.99   \n",
      "...                      ...                    ...        ...        ...   \n",
      "14331               12.34800               1.212660  -0.010415       3.82   \n",
      "14332              -14.29380               2.771240  -0.026277       3.82   \n",
      "14333              -13.20730               3.822240   0.018446       3.82   \n",
      "14334              -11.09120               3.026110  -0.018991       3.82   \n",
      "14335               39.39830             -40.971700  -0.000814       5.91   \n",
      "\n",
      "       trv  CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0      0.0  2.845761                    1.350          0.0               1.0   \n",
      "1      0.0  2.845761                    1.350          0.0               1.0   \n",
      "2      0.0  2.845761                    1.350          0.0               1.0   \n",
      "3      0.0  2.845761                    1.350          0.0               1.0   \n",
      "4      0.0  2.845761                    1.350          0.0               1.0   \n",
      "...    ...       ...                      ...          ...               ...   \n",
      "14331  0.0  3.770894                    2.238          0.0               1.0   \n",
      "14332  0.0  3.770894                    2.238          0.0               1.0   \n",
      "14333  0.0  3.770894                    2.238          0.0               1.0   \n",
      "14334  0.0  3.770894                    2.238          0.0               1.0   \n",
      "14335  0.0  5.271873                    9.457          0.0               0.0   \n",
      "\n",
      "       synonymous_variant    hazard    ClinicalSignificance  \n",
      "0                     0.0  0.628929  Uncertain significance  \n",
      "1                     0.0  0.377715  Uncertain significance  \n",
      "2                     0.0  0.977332  Uncertain significance  \n",
      "3                     0.0  0.360977  Uncertain significance  \n",
      "4                     0.0  0.283255  Uncertain significance  \n",
      "...                   ...       ...                     ...  \n",
      "14331                 0.0  1.597223  Uncertain significance  \n",
      "14332                 0.0  0.738357  Uncertain significance  \n",
      "14333                 0.0  1.666538  Uncertain significance  \n",
      "14334                 0.0  0.470304  Uncertain significance  \n",
      "14335                 1.0  4.587587                     NaN  \n",
      "\n",
      "[14336 rows x 17 columns]\n",
      "ALPK3\n",
      "       sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0      0.0            True   False  80.032854  80.032854   \n",
      "1      1.0           False   False  77.702943  77.702943   \n",
      "2      1.0           False   False  61.368925  61.368925   \n",
      "3      1.0            True   False  82.866530  82.866530   \n",
      "4      0.0            True   False  74.866530  74.866530   \n",
      "...    ...             ...     ...        ...        ...   \n",
      "16591  0.0           False   False  65.199179  65.199179   \n",
      "16592  1.0           False   False  78.300000  78.300000   \n",
      "16593  1.0           False   False  79.537303  79.537303   \n",
      "16594  1.0            True   False  75.950719  75.950719   \n",
      "16595  1.0           False   False  74.951403  74.951403   \n",
      "\n",
      "       principal_component_1  principal_component_4  prs_score  GERP++_RS  \\\n",
      "0                   -11.1928               6.432880  -0.014985       5.97   \n",
      "1                   -13.6237               1.445050  -0.005863       5.97   \n",
      "2                   -10.7004              -0.429001  -0.012379       5.97   \n",
      "3                   -12.9627               0.040513   0.006054       5.97   \n",
      "4                   -14.0718               0.958428   0.011737       5.97   \n",
      "...                      ...                    ...        ...        ...   \n",
      "16591               152.3560             -14.283200  -0.016554       5.97   \n",
      "16592               166.5870              -7.611840   0.017814       5.97   \n",
      "16593               158.8740             -14.447600  -0.018105       5.97   \n",
      "16594               158.7080             -17.049700  -0.029873       5.97   \n",
      "16595               107.2860              34.802200  -0.003167       5.97   \n",
      "\n",
      "       trv  CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0      1.0  4.831172                   10.003          0.0               0.0   \n",
      "1      1.0  4.831172                   10.003          0.0               0.0   \n",
      "2      1.0  4.831172                   10.003          0.0               0.0   \n",
      "3      1.0  4.831172                   10.003          0.0               0.0   \n",
      "4      1.0  4.831172                   10.003          0.0               0.0   \n",
      "...    ...       ...                      ...          ...               ...   \n",
      "16591  0.0  4.831172                   10.003          0.0               0.0   \n",
      "16592  0.0  4.831172                   10.003          0.0               0.0   \n",
      "16593  0.0  4.831172                   10.003          0.0               0.0   \n",
      "16594  0.0  4.831172                   10.003          0.0               0.0   \n",
      "16595  0.0  4.831172                   10.003          0.0               0.0   \n",
      "\n",
      "       synonymous_variant    hazard  ClinicalSignificance  \n",
      "0                     1.0  0.197410         Likely benign  \n",
      "1                     1.0  0.468384         Likely benign  \n",
      "2                     1.0  1.194870         Likely benign  \n",
      "3                     1.0  0.571921         Likely benign  \n",
      "4                     1.0  0.662441         Likely benign  \n",
      "...                   ...       ...                   ...  \n",
      "16591                 1.0  0.296269  Benign/Likely benign  \n",
      "16592                 1.0  0.798722  Benign/Likely benign  \n",
      "16593                 1.0  0.152802  Benign/Likely benign  \n",
      "16594                 1.0  0.153786  Benign/Likely benign  \n",
      "16595                 1.0  0.619004                   NaN  \n",
      "\n",
      "[16596 rows x 17 columns]\n",
      "FLNC\n",
      "       sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0      0.0           False   False  76.199863  76.199863   \n",
      "1      0.0            True   False  79.452430  79.452430   \n",
      "2      0.0           False   False  61.779603  61.779603   \n",
      "3      1.0            True   False  55.619439  55.619439   \n",
      "4      0.0           False   False  62.300000  62.300000   \n",
      "...    ...             ...     ...        ...        ...   \n",
      "10873  0.0           False   False  57.451061  57.451061   \n",
      "10874  1.0            True   False  76.451745  76.451745   \n",
      "10875  0.0            True   False  64.780287  64.780287   \n",
      "10876  1.0           False   False  78.535250  78.535250   \n",
      "10877  0.0            True   False  69.702943  69.702943   \n",
      "\n",
      "       principal_component_1  principal_component_4  prs_score  GERP++_RS  \\\n",
      "0                    4.80640              -29.66200  -0.017680       5.36   \n",
      "1                  206.89500              -25.18670   0.002478       5.46   \n",
      "2                  352.93500                7.02098   0.015626       5.46   \n",
      "3                  -15.14410                1.57492  -0.019519       5.46   \n",
      "4                  -10.87970                1.41830   0.002170       5.46   \n",
      "...                      ...                    ...        ...        ...   \n",
      "10873               -3.62662              -17.94750   0.001124      -1.08   \n",
      "10874               29.21390              -54.08960  -0.012223      -1.08   \n",
      "10875               66.87180               11.34240   0.002801      -1.08   \n",
      "10876               68.11140               15.44270   0.006023      -1.08   \n",
      "10877               35.87750              -50.14360  -0.012927      -1.08   \n",
      "\n",
      "       trv  CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0      0.0  3.911669                    1.839          1.0               0.0   \n",
      "1      0.0  3.932996                    4.718          0.0               1.0   \n",
      "2      1.0  2.785808                    5.947          0.0               1.0   \n",
      "3      0.0  2.970605                    5.799          0.0               1.0   \n",
      "4      0.0  2.970605                    5.799          0.0               1.0   \n",
      "...    ...       ...                      ...          ...               ...   \n",
      "10873  1.0 -0.089889                   -0.424          0.0               1.0   \n",
      "10874  1.0 -0.089889                   -0.424          0.0               1.0   \n",
      "10875  1.0 -0.089889                   -0.424          0.0               1.0   \n",
      "10876  1.0 -0.089889                   -0.424          0.0               1.0   \n",
      "10877  1.0 -0.089889                   -0.424          0.0               1.0   \n",
      "\n",
      "       synonymous_variant    hazard    ClinicalSignificance  \n",
      "0                     0.0  2.561144                     NaN  \n",
      "1                     0.0  0.032574  Uncertain significance  \n",
      "2                     0.0  0.017916                     NaN  \n",
      "3                     0.0  7.065743  Uncertain significance  \n",
      "4                     0.0  1.450566  Uncertain significance  \n",
      "...                   ...       ...                     ...  \n",
      "10873                 0.0  5.553243  Uncertain significance  \n",
      "10874                 0.0  5.190528  Uncertain significance  \n",
      "10875                 0.0  0.814332  Uncertain significance  \n",
      "10876                 0.0  0.229746  Uncertain significance  \n",
      "10877                 0.0  2.576638  Uncertain significance  \n",
      "\n",
      "[10878 rows x 17 columns]\n",
      "MYBPC3\n",
      "       sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0      0.0            True   False  60.500000  60.500000   \n",
      "1      1.0            True   False  67.033539  67.033539   \n",
      "2      0.0            True   False  74.702259  74.702259   \n",
      "3      1.0            True   False  62.866530  62.866530   \n",
      "4      0.0            True   False  78.781656  78.781656   \n",
      "...    ...             ...     ...        ...        ...   \n",
      "13614  0.0           False   False  67.865845  67.865845   \n",
      "13615  1.0           False   False  54.031485  54.031485   \n",
      "13616  1.0           False   False  73.864476  73.864476   \n",
      "13617  0.0            True   False  82.031485  82.031485   \n",
      "13618  0.0            True   False  72.284736  72.284736   \n",
      "\n",
      "       principal_component_1  principal_component_4  prs_score  GERP++_RS  \\\n",
      "0                  -11.13280              -0.622067   0.001330       5.33   \n",
      "1                  -12.49660              -0.604222  -0.020890       5.33   \n",
      "2                    5.63861             -48.776900  -0.000332       5.33   \n",
      "3                  -13.23630              -0.836808   0.012845       5.33   \n",
      "4                  -11.52220               1.923550  -0.009238       5.33   \n",
      "...                      ...                    ...        ...        ...   \n",
      "13614              -14.14200               2.892010  -0.030614       3.53   \n",
      "13615              -10.73910              -4.201490  -0.004927       3.53   \n",
      "13616              -12.02240              -2.800360   0.000215       5.33   \n",
      "13617              275.48000              -0.984977   0.016019       5.33   \n",
      "13618              -11.34290               5.520530   0.011343       1.80   \n",
      "\n",
      "       trv  CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0      0.0  2.419307                    2.345          0.0               1.0   \n",
      "1      0.0  2.419307                    2.345          0.0               1.0   \n",
      "2      0.0  2.419307                    2.345          0.0               1.0   \n",
      "3      0.0  2.419307                    2.345          0.0               1.0   \n",
      "4      0.0  2.419307                    2.345          0.0               1.0   \n",
      "...    ...       ...                      ...          ...               ...   \n",
      "13614  0.0  4.223968                    3.027          0.0               1.0   \n",
      "13615  0.0  4.223968                    3.027          0.0               1.0   \n",
      "13616  0.0  4.818941                    9.994          0.0               0.0   \n",
      "13617  0.0  4.818941                    9.994          0.0               0.0   \n",
      "13618  1.0  0.767623                    0.029          0.0               1.0   \n",
      "\n",
      "       synonymous_variant     hazard    ClinicalSignificance  \n",
      "0                     0.0   3.323108  Uncertain significance  \n",
      "1                     0.0   0.871218  Uncertain significance  \n",
      "2                     0.0   0.303992  Uncertain significance  \n",
      "3                     0.0  13.436711  Uncertain significance  \n",
      "4                     0.0   0.161852  Uncertain significance  \n",
      "...                   ...        ...                     ...  \n",
      "13614                 0.0   0.095230  Uncertain significance  \n",
      "13615                 0.0   3.409022  Uncertain significance  \n",
      "13616                 1.0   1.267508                     NaN  \n",
      "13617                 1.0   7.687717                     NaN  \n",
      "13618                 0.0   6.169122                     NaN  \n",
      "\n",
      "[13619 rows x 17 columns]\n",
      "MYH6\n",
      "       sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0      1.0           False   False  68.618754  68.618754   \n",
      "1      1.0            True   False  66.283368  66.283368   \n",
      "2      1.0            True   False  67.370294  67.370294   \n",
      "3      0.0            True   False  68.703628  68.703628   \n",
      "4      0.0            True   False  58.951403  58.951403   \n",
      "...    ...             ...     ...        ...        ...   \n",
      "11017  1.0           False   False  71.600000  71.600000   \n",
      "11018  0.0           False   False  81.032170  81.032170   \n",
      "11019  0.0           False   False  73.702943  73.702943   \n",
      "11020  0.0           False   False  58.116359  58.116359   \n",
      "11021  1.0           False    True  75.236140  77.618070   \n",
      "\n",
      "       principal_component_1  principal_component_4  prs_score  GERP++_RS  \\\n",
      "0                  -11.92330               4.296090  -0.021392       4.69   \n",
      "1                  -13.07810               2.636580  -0.009047       4.69   \n",
      "2                  -13.76660              -2.825800   0.005554       4.69   \n",
      "3                  255.10600              -0.243969   0.005543       4.69   \n",
      "4                   72.93780               1.925720   0.013355       4.69   \n",
      "...                      ...                    ...        ...        ...   \n",
      "11017              -17.04400               6.027940  -0.001154       4.11   \n",
      "11018              -13.94500              -1.182350   0.007774       4.11   \n",
      "11019              -14.66560               9.950640  -0.020078       4.11   \n",
      "11020              -13.57150               8.237130  -0.021659       4.11   \n",
      "11021               -9.30501               0.320686  -0.001007       4.11   \n",
      "\n",
      "       trv  CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0      0.0  1.578815                    0.835          0.0               1.0   \n",
      "1      0.0  1.578815                    0.835          0.0               1.0   \n",
      "2      0.0  1.578815                    0.835          0.0               1.0   \n",
      "3      0.0  1.578815                    0.835          0.0               1.0   \n",
      "4      0.0  1.578815                    0.835          0.0               1.0   \n",
      "...    ...       ...                      ...          ...               ...   \n",
      "11017  1.0  2.515735                    6.157          0.0               1.0   \n",
      "11018  1.0  2.515735                    6.157          0.0               1.0   \n",
      "11019  1.0  2.515735                    6.157          0.0               1.0   \n",
      "11020  1.0  2.515735                    6.157          0.0               1.0   \n",
      "11021  1.0  2.515735                    6.157          0.0               1.0   \n",
      "\n",
      "       synonymous_variant     hazard    ClinicalSignificance  \n",
      "0                     0.0   2.024724  Uncertain significance  \n",
      "1                     0.0   9.955542  Uncertain significance  \n",
      "2                     0.0  15.386250  Uncertain significance  \n",
      "3                     0.0   3.603002  Uncertain significance  \n",
      "4                     0.0  16.243633  Uncertain significance  \n",
      "...                   ...        ...                     ...  \n",
      "11017                 0.0   4.623007  Uncertain significance  \n",
      "11018                 0.0   3.319198  Uncertain significance  \n",
      "11019                 0.0   1.586927  Uncertain significance  \n",
      "11020                 0.0   2.407756  Uncertain significance  \n",
      "11021                 0.0   3.566494  Uncertain significance  \n",
      "\n",
      "[11022 rows x 17 columns]\n",
      "MYH7\n",
      "      sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0     0.0           False   False  68.284736  68.284736   \n",
      "1     0.0           False   False  67.780972  67.780972   \n",
      "2     1.0            True   False  71.370294  71.370294   \n",
      "3     1.0           False   False  56.780287  56.780287   \n",
      "4     1.0           False   False  58.617385  58.617385   \n",
      "...   ...             ...     ...        ...        ...   \n",
      "3133  1.0            True   False  67.200548  67.200548   \n",
      "3134  0.0            True   False  66.200000  66.200000   \n",
      "3135  0.0           False   False  61.864476  61.864476   \n",
      "3136  0.0            True   False  75.950719  75.950719   \n",
      "3137  1.0            True   False  65.117043  65.117043   \n",
      "\n",
      "      principal_component_1  principal_component_4  prs_score  GERP++_RS  trv  \\\n",
      "0                  -10.4357               0.900290  -0.011304       6.05  0.0   \n",
      "1                   48.7103              11.805900   0.004079       6.05  0.0   \n",
      "2                  -10.1565               3.559080   0.005588       5.93  0.0   \n",
      "3                  -11.6771              -0.692553  -0.006299       5.93  1.0   \n",
      "4                  149.6580             -10.230000   0.000874       6.05  0.0   \n",
      "...                     ...                    ...        ...        ...  ...   \n",
      "3133                66.8029               5.722700  -0.000097       5.80  1.0   \n",
      "3134               -13.3948               2.835480  -0.013630       5.80  0.0   \n",
      "3135               -12.2916               7.027540  -0.013844       5.80  0.0   \n",
      "3136               -13.8181               1.073980   0.032964       5.80  0.0   \n",
      "3137               -12.4674              -1.179830  -0.007558       5.80  0.0   \n",
      "\n",
      "      CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0     4.578059                    9.994          0.0               0.0   \n",
      "1     4.578059                    9.994          0.0               0.0   \n",
      "2     3.846308                    7.900          0.0               1.0   \n",
      "3     2.569327                    7.218          0.0               1.0   \n",
      "4     4.578059                    9.994          0.0               0.0   \n",
      "...        ...                      ...          ...               ...   \n",
      "3133  3.023692                    6.475          0.0               1.0   \n",
      "3134  3.714583                    8.150          0.0               1.0   \n",
      "3135  3.714583                    8.150          0.0               1.0   \n",
      "3136  3.714583                    8.150          0.0               1.0   \n",
      "3137  3.714583                    8.150          0.0               1.0   \n",
      "\n",
      "      synonymous_variant     hazard    ClinicalSignificance  \n",
      "0                    1.0   0.529407                     NaN  \n",
      "1                    1.0   2.077006                     NaN  \n",
      "2                    0.0  25.808316                     NaN  \n",
      "3                    0.0   0.000132  Uncertain significance  \n",
      "4                    1.0  10.090340           Likely benign  \n",
      "...                  ...        ...                     ...  \n",
      "3133                 0.0   0.002980                     NaN  \n",
      "3134                 0.0  10.851368  Uncertain significance  \n",
      "3135                 0.0   4.564981  Uncertain significance  \n",
      "3136                 0.0  12.301288  Uncertain significance  \n",
      "3137                 0.0  30.067327  Uncertain significance  \n",
      "\n",
      "[3138 rows x 17 columns]\n",
      "PTPN11\n",
      "     sex  is_family_hist  is_hcm   duration        age  principal_component_1  \\\n",
      "0    1.0           False   False  59.865845  59.865845               -11.4414   \n",
      "1    0.0           False   False  58.900000  58.900000               -11.3896   \n",
      "2    1.0           False   False  83.537303  83.537303               -12.9004   \n",
      "3    1.0           False   False  79.950719  79.950719               -13.9416   \n",
      "4    0.0           False   False  63.537303  63.537303               387.8660   \n",
      "..   ...             ...     ...        ...        ...                    ...   \n",
      "623  1.0            True   False  82.368241  82.368241               -13.5266   \n",
      "624  1.0            True   False  59.118412  59.118412               -13.7848   \n",
      "625  0.0           False   False  75.780972  75.780972               -12.1423   \n",
      "626  1.0            True   False  70.617385  70.617385                83.8745   \n",
      "627  0.0           False   False  69.284052  69.284052                87.0210   \n",
      "\n",
      "     principal_component_4  prs_score  GERP++_RS  trv  CADD_raw  \\\n",
      "0                -1.199300   0.019187       4.63  0.0  3.848541   \n",
      "1                -3.054470  -0.012068       4.63  0.0  3.848541   \n",
      "2                -0.856772   0.001376       4.63  0.0  3.848541   \n",
      "3                 0.387402  -0.012135       4.63  0.0  3.848541   \n",
      "4                 4.702810  -0.004294       4.46  0.0  3.971822   \n",
      "..                     ...        ...        ...  ...       ...   \n",
      "623               3.106260   0.009650      -7.71  1.0  3.926316   \n",
      "624               1.820340  -0.021000      -7.71  1.0  3.926316   \n",
      "625               3.481250  -0.005707      -7.71  1.0  3.926316   \n",
      "626              25.058200   0.015649       4.41  1.0  4.001367   \n",
      "627              25.126400   0.007869       4.41  1.0  4.001367   \n",
      "\n",
      "     phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0                      5.318          0.0               1.0   \n",
      "1                      5.318          0.0               1.0   \n",
      "2                      5.318          0.0               1.0   \n",
      "3                      5.318          0.0               1.0   \n",
      "4                      0.944          0.0               1.0   \n",
      "..                       ...          ...               ...   \n",
      "623                   -2.267          1.0               0.0   \n",
      "624                   -2.267          1.0               0.0   \n",
      "625                   -2.267          1.0               0.0   \n",
      "626                    0.910          0.0               1.0   \n",
      "627                    0.910          0.0               1.0   \n",
      "\n",
      "     synonymous_variant        hazard  \\\n",
      "0                   0.0  1.493901e+05   \n",
      "1                   0.0  2.542879e+05   \n",
      "2                   0.0  8.658807e+03   \n",
      "3                   0.0  1.833369e+04   \n",
      "4                   0.0  2.987165e-27   \n",
      "..                  ...           ...   \n",
      "623                 0.0  4.835295e-01   \n",
      "624                 0.0  6.995924e-01   \n",
      "625                 0.0  6.296013e-01   \n",
      "626                 0.0  9.122294e-01   \n",
      "627                 0.0  1.092672e+00   \n",
      "\n",
      "                             ClinicalSignificance  \n",
      "0                                             NaN  \n",
      "1                                             NaN  \n",
      "2                                             NaN  \n",
      "3                                             NaN  \n",
      "4    Conflicting interpretations of pathogenicity  \n",
      "..                                            ...  \n",
      "623                                           NaN  \n",
      "624                                           NaN  \n",
      "625                                           NaN  \n",
      "626                        Uncertain significance  \n",
      "627                        Uncertain significance  \n",
      "\n",
      "[628 rows x 17 columns]\n",
      "TNNI3\n",
      "      sex  is_family_hist  is_hcm   duration        age  \\\n",
      "0     0.0            True   False  70.866530  70.866530   \n",
      "1     1.0           False   False  63.619439  63.619439   \n",
      "2     0.0            True   False  79.704312  79.704312   \n",
      "3     1.0           False   False  72.339493  73.368925   \n",
      "4     0.0           False   False  74.031485  74.031485   \n",
      "...   ...             ...     ...        ...        ...   \n",
      "1019  0.0            True   False  76.451745  76.451745   \n",
      "1020  1.0           False   False  73.779603  73.779603   \n",
      "1021  1.0            True   False  73.702943  73.702943   \n",
      "1022  0.0           False   False  61.779603  61.779603   \n",
      "1023  0.0           False   False  75.285421  75.285421   \n",
      "\n",
      "      principal_component_1  principal_component_4  prs_score  GERP++_RS  trv  \\\n",
      "0                 -14.29550               4.279050   0.002829       1.35  0.0   \n",
      "1                  -8.91178              -1.432520  -0.020134      -5.10  0.0   \n",
      "2                 -11.74950               5.313980  -0.005804      -5.10  0.0   \n",
      "3                 -11.65980               1.294020  -0.022743      -5.10  0.0   \n",
      "4                  -4.81544             -20.116900  -0.004452      -5.10  0.0   \n",
      "...                     ...                    ...        ...        ...  ...   \n",
      "1019              -15.91160              -2.433730  -0.003453       6.08  0.0   \n",
      "1020              -11.49580              -3.183370   0.019623       6.08  0.0   \n",
      "1021              -10.52150              -1.133010  -0.007748       6.08  0.0   \n",
      "1022              -10.40940               0.300187   0.000272       6.08  0.0   \n",
      "1023              -11.27910              -2.340820  -0.006197       6.08  0.0   \n",
      "\n",
      "      CADD_raw  phyloP100way_vertebrate  deleterious  missense_variant  \\\n",
      "0     1.835163                    0.796          0.0               1.0   \n",
      "1    -0.153535                   -0.437          0.0               1.0   \n",
      "2    -0.153535                   -0.437          0.0               1.0   \n",
      "3    -0.153535                   -0.437          0.0               1.0   \n",
      "4    -0.153535                   -0.437          0.0               1.0   \n",
      "...        ...                      ...          ...               ...   \n",
      "1019  4.316262                    7.679          0.0               0.0   \n",
      "1020  4.316262                    7.679          0.0               0.0   \n",
      "1021  4.316262                    7.679          0.0               0.0   \n",
      "1022  4.316262                    7.679          0.0               0.0   \n",
      "1023  4.316262                    7.679          0.0               0.0   \n",
      "\n",
      "      synonymous_variant         hazard  \\\n",
      "0                    0.0    4842.679020   \n",
      "1                    0.0  236644.989810   \n",
      "2                    0.0    2946.463182   \n",
      "3                    0.0   99059.724867   \n",
      "4                    0.0       0.706390   \n",
      "...                  ...            ...   \n",
      "1019                 1.0       8.099429   \n",
      "1020                 1.0       1.657433   \n",
      "1021                 1.0     206.184234   \n",
      "1022                 1.0       0.574741   \n",
      "1023                 1.0       0.422488   \n",
      "\n",
      "                              ClinicalSignificance  \n",
      "0                           Uncertain significance  \n",
      "1     Conflicting interpretations of pathogenicity  \n",
      "2     Conflicting interpretations of pathogenicity  \n",
      "3     Conflicting interpretations of pathogenicity  \n",
      "4     Conflicting interpretations of pathogenicity  \n",
      "...                                            ...  \n",
      "1019                                 Likely benign  \n",
      "1020                                 Likely benign  \n",
      "1021                                 Likely benign  \n",
      "1022                                 Likely benign  \n",
      "1023                                 Likely benign  \n",
      "\n",
      "[1024 rows x 17 columns]\n",
      "TTR\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P-value</th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>Thresholds</th>\n",
       "      <th>F1 scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACTN2</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4.444509e+06</td>\n",
       "      <td>2</td>\n",
       "      <td>0.002151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALPK3</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.224834e+07</td>\n",
       "      <td>2</td>\n",
       "      <td>0.020339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNC</th>\n",
       "      <td>9.406915e-07</td>\n",
       "      <td>6.811688e+00</td>\n",
       "      <td>83</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYBPC3</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.901183e+07</td>\n",
       "      <td>11</td>\n",
       "      <td>0.144796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH6</th>\n",
       "      <td>1.676823e-07</td>\n",
       "      <td>4.465473e+01</td>\n",
       "      <td>99</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH7</th>\n",
       "      <td>2.688528e-06</td>\n",
       "      <td>1.490769e+01</td>\n",
       "      <td>99</td>\n",
       "      <td>0.057762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPN11</th>\n",
       "      <td>2.679943e-01</td>\n",
       "      <td>1.025288e+01</td>\n",
       "      <td>54</td>\n",
       "      <td>0.150510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNNI3</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5.215107e+06</td>\n",
       "      <td>14</td>\n",
       "      <td>0.523810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTR</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.478141e+06</td>\n",
       "      <td>28</td>\n",
       "      <td>0.629442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             P-value    Odds ratio  Thresholds  F1 scores\n",
       "ACTN2   1.000000e+00  4.444509e+06           2   0.002151\n",
       "ALPK3   1.000000e+00  1.224834e+07           2   0.020339\n",
       "FLNC    9.406915e-07  6.811688e+00          83   0.000000\n",
       "MYBPC3  1.000000e+00  3.901183e+07          11   0.144796\n",
       "MYH6    1.676823e-07  4.465473e+01          99   0.000000\n",
       "MYH7    2.688528e-06  1.490769e+01          99   0.057762\n",
       "PTPN11  2.679943e-01  1.025288e+01          54   0.150510\n",
       "TNNI3   1.000000e+00  5.215107e+06          14   0.523810\n",
       "TTR     1.000000e+00  6.478141e+06          28   0.629442"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "updated_p_dict, hazard_ratios, f1_scores = find_threshold_vals(threshold_by_odds_ratio)\n",
    "p_vals = pd.DataFrame.from_dict(updated_p_dict, orient = 'index')\n",
    "p_vals.columns = [\"P-value\"]\n",
    "hazard_ratios = pd.DataFrame.from_dict(hazard_ratios, orient = 'index')\n",
    "hazard_ratios.columns = [\"Odds ratio\"]\n",
    "thresholds = pd.DataFrame.from_dict(threshold_by_odds_ratio, orient = 'index')\n",
    "thresholds.columns = [\"Thresholds\"]\n",
    "f1_scores = pd.DataFrame.from_dict(f1_scores, orient = 'index')\n",
    "f1_scores.columns = [\"F1 scores\"]\n",
    "df = p_vals.join(hazard_ratios).join(thresholds).join(f1_scores)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert highest F1 scores to dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTN2\n",
      "ALPK3\n",
      "FLNC\n",
      "MYBPC3\n",
      "MYH6\n",
      "MYH7\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[42], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m updated_p_dict, hazard_ratios, f1_scores \u001b[38;5;241m=\u001b[39m \u001b[43mfind_threshold_vals\u001b[49m\u001b[43m(\u001b[49m\u001b[43mthresholds_by_f1_score\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m p_vals \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame\u001b[38;5;241m.\u001b[39mfrom_dict(updated_p_dict, orient \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      3\u001b[0m p_vals\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mP-value\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "Cell \u001b[0;32mIn[34], line 14\u001b[0m, in \u001b[0;36mfind_threshold_vals\u001b[0;34m(dict)\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mdict\u001b[39m:\n\u001b[1;32m     11\u001b[0m         threshold \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m[gene]\n\u001b[0;32m---> 14\u001b[0m         lifelines_data \u001b[38;5;241m=\u001b[39m \u001b[43mcross_val\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     18\u001b[0m         percentiles \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mpercentile(lifelines_data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhazard\u001b[39m\u001b[38;5;124m'\u001b[39m], [threshold])\n\u001b[1;32m     19\u001b[0m         bottom \u001b[38;5;241m=\u001b[39m lifelines_data[lifelines_data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhazard\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m<\u001b[39m percentiles[\u001b[38;5;241m0\u001b[39m]]\n",
      "Cell \u001b[0;32mIn[30], line 62\u001b[0m, in \u001b[0;36mcross_val\u001b[0;34m(gene)\u001b[0m\n\u001b[1;32m     57\u001b[0m test \u001b[38;5;241m=\u001b[39m test\u001b[38;5;241m.\u001b[39mdrop([\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mvar_index\u001b[39m\u001b[38;5;124m'\u001b[39m], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m     59\u001b[0m \u001b[38;5;66;03m# fit CPH and add hazard scores\u001b[39;00m\n\u001b[0;32m---> 62\u001b[0m \u001b[43mcph\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrain\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mduration_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mduration\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mevent_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mis_hcm\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfit_options\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstep_size\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[38;5;241;43m0.1\u001b[39;49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     63\u001b[0m hazard_scores_fold \u001b[38;5;241m=\u001b[39m cph\u001b[38;5;241m.\u001b[39mpredict_partial_hazard(test)\n\u001b[1;32m     64\u001b[0m test[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhazard\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m hazard_scores_fold\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/utils/__init__.py:56\u001b[0m, in \u001b[0;36mCensoringType.right_censoring.<locals>.f\u001b[0;34m(model, *args, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(function)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m(model, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mset_censoring_type(model, \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mRIGHT)\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:290\u001b[0m, in \u001b[0;36mCoxPHFitter.fit\u001b[0;34m(self, df, duration_col, event_col, show_progress, initial_point, strata, weights_col, cluster_col, robust, batch_mode, timeline, formula, entry_col, fit_options)\u001b[0m\n\u001b[1;32m    184\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    185\u001b[0m \u001b[38;5;124;03mFit the Cox proportional hazard model to a right-censored dataset. Alias of `fit_right_censoring`.\u001b[39;00m\n\u001b[1;32m    186\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    287\u001b[0m \n\u001b[1;32m    288\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    289\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata \u001b[38;5;241m=\u001b[39m utils\u001b[38;5;241m.\u001b[39m_to_list_or_singleton(utils\u001b[38;5;241m.\u001b[39mcoalesce(strata, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata))\n\u001b[0;32m--> 290\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_model \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    291\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    292\u001b[0m \u001b[43m    \u001b[49m\u001b[43mduration_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    293\u001b[0m \u001b[43m    \u001b[49m\u001b[43mevent_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mevent_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    294\u001b[0m \u001b[43m    \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mshow_progress\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    295\u001b[0m \u001b[43m    \u001b[49m\u001b[43minitial_point\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_point\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    296\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstrata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstrata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    297\u001b[0m \u001b[43m    \u001b[49m\u001b[43mweights_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mweights_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    298\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcluster_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcluster_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    299\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrobust\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrobust\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    300\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbatch_mode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_mode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    301\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeline\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    302\u001b[0m \u001b[43m    \u001b[49m\u001b[43mformula\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mformula\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    303\u001b[0m \u001b[43m    \u001b[49m\u001b[43mentry_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mentry_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    304\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfit_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfit_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    305\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    306\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:610\u001b[0m, in \u001b[0;36mCoxPHFitter._fit_model\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    608\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_fit_model\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    609\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbaseline_estimation_method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbreslow\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m--> 610\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model_breslow\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    611\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbaseline_estimation_method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mspline\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    612\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fit_model_spline(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:623\u001b[0m, in \u001b[0;36mCoxPHFitter._fit_model_breslow\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    619\u001b[0m model \u001b[38;5;241m=\u001b[39m SemiParametricPHFitter(\n\u001b[1;32m    620\u001b[0m     penalizer\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpenalizer, l1_ratio\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39ml1_ratio, strata\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata, alpha\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39malpha, label\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_label\n\u001b[1;32m    621\u001b[0m )\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m utils\u001b[38;5;241m.\u001b[39mCensoringType\u001b[38;5;241m.\u001b[39mis_right_censoring(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 623\u001b[0m     \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    624\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m model\n\u001b[1;32m    625\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/utils/__init__.py:56\u001b[0m, in \u001b[0;36mCensoringType.right_censoring.<locals>.f\u001b[0;34m(model, *args, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(function)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m(model, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mset_censoring_type(model, \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mRIGHT)\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1242\u001b[0m, in \u001b[0;36mSemiParametricPHFitter.fit\u001b[0;34m(self, df, duration_col, event_col, show_progress, initial_point, strata, weights_col, cluster_col, robust, batch_mode, timeline, formula, entry_col, fit_options)\u001b[0m\n\u001b[1;32m   1239\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mweights\u001b[38;5;241m.\u001b[39mindex \u001b[38;5;241m=\u001b[39m original_index\n\u001b[1;32m   1241\u001b[0m \u001b[38;5;66;03m# TODO: doesn't handle weights, nor strata\u001b[39;00m\n\u001b[0;32m-> 1242\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_central_values \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_compute_central_values_of_raw_training_data\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstrata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1244\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_norm_mean \u001b[38;5;241m=\u001b[39m X\u001b[38;5;241m.\u001b[39mmean(\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m   1245\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_norm_std \u001b[38;5;241m=\u001b[39m X\u001b[38;5;241m.\u001b[39mstd(\u001b[38;5;241m0\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/lifelines/fitters/__init__.py:1297\u001b[0m, in \u001b[0;36mRegressionFitter._compute_central_values_of_raw_training_data\u001b[0;34m(self, df, strata, name)\u001b[0m\n\u001b[1;32m   1295\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1296\u001b[0m     describe_kwarg \u001b[38;5;241m=\u001b[39m {}\n\u001b[0;32m-> 1297\u001b[0m described \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdescribe\u001b[49m\u001b[43m(\u001b[49m\u001b[43minclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mall\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mdescribe_kwarg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1298\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m described\u001b[38;5;241m.\u001b[39mindex \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m50\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m described\u001b[38;5;241m.\u001b[39mindex:\n\u001b[1;32m   1299\u001b[0m     central_stats \u001b[38;5;241m=\u001b[39m described\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mcopy()\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/generic.py:11544\u001b[0m, in \u001b[0;36mNDFrame.describe\u001b[0;34m(self, percentiles, include, exclude)\u001b[0m\n\u001b[1;32m  11302\u001b[0m \u001b[38;5;129m@final\u001b[39m\n\u001b[1;32m  11303\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdescribe\u001b[39m(\n\u001b[1;32m  11304\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  11307\u001b[0m     exclude\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m  11308\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Self:\n\u001b[1;32m  11309\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m  11310\u001b[0m \u001b[38;5;124;03m    Generate descriptive statistics.\u001b[39;00m\n\u001b[1;32m  11311\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  11542\u001b[0m \u001b[38;5;124;03m    max            NaN      3.0\u001b[39;00m\n\u001b[1;32m  11543\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m> 11544\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdescribe_ndframe\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m  11545\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m  11546\u001b[0m \u001b[43m        \u001b[49m\u001b[43minclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minclude\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  11547\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexclude\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  11548\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpercentiles\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpercentiles\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  11549\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdescribe\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/methods/describe.py:97\u001b[0m, in \u001b[0;36mdescribe_ndframe\u001b[0;34m(obj, include, exclude, percentiles)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     91\u001b[0m     describer \u001b[38;5;241m=\u001b[39m DataFrameDescriber(\n\u001b[1;32m     92\u001b[0m         obj\u001b[38;5;241m=\u001b[39mcast(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDataFrame\u001b[39m\u001b[38;5;124m\"\u001b[39m, obj),\n\u001b[1;32m     93\u001b[0m         include\u001b[38;5;241m=\u001b[39minclude,\n\u001b[1;32m     94\u001b[0m         exclude\u001b[38;5;241m=\u001b[39mexclude,\n\u001b[1;32m     95\u001b[0m     )\n\u001b[0;32m---> 97\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mdescriber\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdescribe\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpercentiles\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpercentiles\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     98\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m cast(NDFrameT, result)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/methods/describe.py:170\u001b[0m, in \u001b[0;36mDataFrameDescriber.describe\u001b[0;34m(self, percentiles)\u001b[0m\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m _, series \u001b[38;5;129;01min\u001b[39;00m data\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m    169\u001b[0m     describe_func \u001b[38;5;241m=\u001b[39m select_describe_func(series)\n\u001b[0;32m--> 170\u001b[0m     ldesc\u001b[38;5;241m.\u001b[39mappend(\u001b[43mdescribe_func\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseries\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpercentiles\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    172\u001b[0m col_names \u001b[38;5;241m=\u001b[39m reorder_columns(ldesc)\n\u001b[1;32m    173\u001b[0m d \u001b[38;5;241m=\u001b[39m concat(\n\u001b[1;32m    174\u001b[0m     [x\u001b[38;5;241m.\u001b[39mreindex(col_names, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m ldesc],\n\u001b[1;32m    175\u001b[0m     axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m,\n\u001b[1;32m    176\u001b[0m     sort\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    177\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/methods/describe.py:232\u001b[0m, in \u001b[0;36mdescribe_numeric_1d\u001b[0;34m(series, percentiles)\u001b[0m\n\u001b[1;32m    227\u001b[0m formatted_percentiles \u001b[38;5;241m=\u001b[39m format_percentiles(percentiles)\n\u001b[1;32m    229\u001b[0m stat_index \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcount\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmean\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstd\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmin\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m+\u001b[39m formatted_percentiles \u001b[38;5;241m+\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    230\u001b[0m d \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    231\u001b[0m     [series\u001b[38;5;241m.\u001b[39mcount(), series\u001b[38;5;241m.\u001b[39mmean(), series\u001b[38;5;241m.\u001b[39mstd(), series\u001b[38;5;241m.\u001b[39mmin()]\n\u001b[0;32m--> 232\u001b[0m     \u001b[38;5;241m+\u001b[39m \u001b[43mseries\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquantile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpercentiles\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[1;32m    233\u001b[0m     \u001b[38;5;241m+\u001b[39m [series\u001b[38;5;241m.\u001b[39mmax()]\n\u001b[1;32m    234\u001b[0m )\n\u001b[1;32m    235\u001b[0m \u001b[38;5;66;03m# GH#48340 - always return float on non-complex numeric data\u001b[39;00m\n\u001b[1;32m    236\u001b[0m dtype: DtypeObj \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/series.py:2769\u001b[0m, in \u001b[0;36mSeries.quantile\u001b[0;34m(self, q, interpolation)\u001b[0m\n\u001b[1;32m   2765\u001b[0m \u001b[38;5;66;03m# We dispatch to DataFrame so that core.internals only has to worry\u001b[39;00m\n\u001b[1;32m   2766\u001b[0m \u001b[38;5;66;03m#  about 2D cases.\u001b[39;00m\n\u001b[1;32m   2767\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mto_frame()\n\u001b[0;32m-> 2769\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquantile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minterpolation\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minterpolation\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnumeric_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m   2770\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m2\u001b[39m:\n\u001b[1;32m   2771\u001b[0m     result \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39miloc[:, \u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/frame.py:11831\u001b[0m, in \u001b[0;36mDataFrame.quantile\u001b[0;34m(self, q, axis, numeric_only, interpolation, method)\u001b[0m\n\u001b[1;32m  11827\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m  11828\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInvalid method: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmethod\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m. Method must be in \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mvalid_method\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m  11829\u001b[0m     )\n\u001b[1;32m  11830\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msingle\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m> 11831\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[43mdata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquantile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mqs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minterpolation\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minterpolation\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m  11832\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtable\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m  11833\u001b[0m     valid_interpolation \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnearest\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlower\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhigher\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/internals/managers.py:1507\u001b[0m, in \u001b[0;36mBlockManager.quantile\u001b[0;34m(self, qs, interpolation)\u001b[0m\n\u001b[1;32m   1504\u001b[0m new_axes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes)\n\u001b[1;32m   1505\u001b[0m new_axes[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m=\u001b[39m Index(qs, dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mfloat64)\n\u001b[0;32m-> 1507\u001b[0m blocks \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m   1508\u001b[0m     blk\u001b[38;5;241m.\u001b[39mquantile(qs\u001b[38;5;241m=\u001b[39mqs, interpolation\u001b[38;5;241m=\u001b[39minterpolation) \u001b[38;5;28;01mfor\u001b[39;00m blk \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblocks\n\u001b[1;32m   1509\u001b[0m ]\n\u001b[1;32m   1511\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)(blocks, new_axes)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/internals/managers.py:1508\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1504\u001b[0m new_axes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes)\n\u001b[1;32m   1505\u001b[0m new_axes[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m=\u001b[39m Index(qs, dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mfloat64)\n\u001b[1;32m   1507\u001b[0m blocks \u001b[38;5;241m=\u001b[39m [\n\u001b[0;32m-> 1508\u001b[0m     \u001b[43mblk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquantile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mqs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mqs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minterpolation\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minterpolation\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m blk \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblocks\n\u001b[1;32m   1509\u001b[0m ]\n\u001b[1;32m   1511\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)(blocks, new_axes)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/internals/blocks.py:1587\u001b[0m, in \u001b[0;36mBlock.quantile\u001b[0;34m(self, qs, interpolation)\u001b[0m\n\u001b[1;32m   1584\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m2\u001b[39m\n\u001b[1;32m   1585\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m is_list_like(qs)  \u001b[38;5;66;03m# caller is responsible for this\u001b[39;00m\n\u001b[0;32m-> 1587\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mquantile_compat\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43masarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43mqs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_values\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minterpolation\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1588\u001b[0m \u001b[38;5;66;03m# ensure_block_shape needed for cases where we start with EA and result\u001b[39;00m\n\u001b[1;32m   1589\u001b[0m \u001b[38;5;66;03m#  is ndarray, e.g. IntegerArray, SparseArray\u001b[39;00m\n\u001b[1;32m   1590\u001b[0m result \u001b[38;5;241m=\u001b[39m ensure_block_shape(result, ndim\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/array_algos/quantile.py:39\u001b[0m, in \u001b[0;36mquantile_compat\u001b[0;34m(values, qs, interpolation)\u001b[0m\n\u001b[1;32m     37\u001b[0m     fill_value \u001b[38;5;241m=\u001b[39m na_value_for_dtype(values\u001b[38;5;241m.\u001b[39mdtype, compat\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m     38\u001b[0m     mask \u001b[38;5;241m=\u001b[39m isna(values)\n\u001b[0;32m---> 39\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mquantile_with_mask\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mqs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minterpolation\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     40\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     41\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m values\u001b[38;5;241m.\u001b[39m_quantile(qs, interpolation)\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/array_algos/quantile.py:97\u001b[0m, in \u001b[0;36mquantile_with_mask\u001b[0;34m(values, mask, fill_value, qs, interpolation)\u001b[0m\n\u001b[1;32m     95\u001b[0m     result \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mrepeat(flat, \u001b[38;5;28mlen\u001b[39m(values))\u001b[38;5;241m.\u001b[39mreshape(\u001b[38;5;28mlen\u001b[39m(values), \u001b[38;5;28mlen\u001b[39m(qs))\n\u001b[1;32m     96\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 97\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43m_nanpercentile\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     98\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     99\u001b[0m \u001b[43m        \u001b[49m\u001b[43mqs\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m100.0\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    100\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfill_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    101\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    102\u001b[0m \u001b[43m        \u001b[49m\u001b[43minterpolation\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minterpolation\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    103\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    105\u001b[0m     result \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(result, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    106\u001b[0m     result \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39mT\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/pandas/core/array_algos/quantile.py:218\u001b[0m, in \u001b[0;36m_nanpercentile\u001b[0;34m(values, qs, na_value, mask, interpolation)\u001b[0m\n\u001b[1;32m    216\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[1;32m    217\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpercentile\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    219\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    220\u001b[0m \u001b[43m        \u001b[49m\u001b[43mqs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    221\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    222\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;66;43;03m# error: No overload variant of \"percentile\" matches argument types\u001b[39;49;00m\n\u001b[1;32m    223\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;66;43;03m# \"ndarray[Any, Any]\", \"ndarray[Any, dtype[floating[_64Bit]]]\",\u001b[39;49;00m\n\u001b[1;32m    224\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;66;43;03m# \"int\", \"Dict[str, str]\"  [call-overload]\u001b[39;49;00m\n\u001b[1;32m    225\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minterpolation\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[call-overload]\u001b[39;49;00m\n\u001b[1;32m    226\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/numpy/lib/function_base.py:4283\u001b[0m, in \u001b[0;36mpercentile\u001b[0;34m(a, q, axis, out, overwrite_input, method, keepdims, interpolation)\u001b[0m\n\u001b[1;32m   4281\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m _quantile_is_valid(q):\n\u001b[1;32m   4282\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPercentiles must be in the range [0, 100]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 4283\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_quantile_unchecked\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   4284\u001b[0m \u001b[43m    \u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moverwrite_input\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeepdims\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/numpy/lib/function_base.py:4555\u001b[0m, in \u001b[0;36m_quantile_unchecked\u001b[0;34m(a, q, axis, out, overwrite_input, method, keepdims)\u001b[0m\n\u001b[1;32m   4547\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_quantile_unchecked\u001b[39m(a,\n\u001b[1;32m   4548\u001b[0m                         q,\n\u001b[1;32m   4549\u001b[0m                         axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4552\u001b[0m                         method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlinear\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   4553\u001b[0m                         keepdims\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[1;32m   4554\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Assumes that q is in [0, 1], and is an ndarray\"\"\"\u001b[39;00m\n\u001b[0;32m-> 4555\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_ureduce\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4556\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mfunc\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_quantile_ureduce_func\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4557\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4558\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mkeepdims\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeepdims\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4559\u001b[0m \u001b[43m                    \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4560\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4561\u001b[0m \u001b[43m                    \u001b[49m\u001b[43moverwrite_input\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moverwrite_input\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4562\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/numpy/lib/function_base.py:3823\u001b[0m, in \u001b[0;36m_ureduce\u001b[0;34m(a, func, keepdims, **kwargs)\u001b[0m\n\u001b[1;32m   3820\u001b[0m             index_out \u001b[38;5;241m=\u001b[39m (\u001b[38;5;241m0\u001b[39m, ) \u001b[38;5;241m*\u001b[39m nd\n\u001b[1;32m   3821\u001b[0m             kwargs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mout\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m out[(\u001b[38;5;28mEllipsis\u001b[39m, ) \u001b[38;5;241m+\u001b[39m index_out]\n\u001b[0;32m-> 3823\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3825\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m out \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   3826\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m out\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/numpy/lib/function_base.py:4721\u001b[0m, in \u001b[0;36m_quantile_ureduce_func\u001b[0;34m(a, q, axis, out, overwrite_input, method)\u001b[0m\n\u001b[1;32m   4719\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   4720\u001b[0m         arr \u001b[38;5;241m=\u001b[39m a\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[0;32m-> 4721\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43m_quantile\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4722\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mquantiles\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4723\u001b[0m \u001b[43m                   \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4724\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4725\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4726\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/numpy/lib/function_base.py:4844\u001b[0m, in \u001b[0;36m_quantile\u001b[0;34m(arr, quantiles, axis, method, out)\u001b[0m\n\u001b[1;32m   4839\u001b[0m     gamma \u001b[38;5;241m=\u001b[39m gamma\u001b[38;5;241m.\u001b[39mreshape(result_shape)\n\u001b[1;32m   4840\u001b[0m     result \u001b[38;5;241m=\u001b[39m _lerp(previous,\n\u001b[1;32m   4841\u001b[0m                    \u001b[38;5;28mnext\u001b[39m,\n\u001b[1;32m   4842\u001b[0m                    gamma,\n\u001b[1;32m   4843\u001b[0m                    out\u001b[38;5;241m=\u001b[39mout)\n\u001b[0;32m-> 4844\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43many\u001b[49m\u001b[43m(\u001b[49m\u001b[43mslices_having_nans\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m   4845\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m out \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   4846\u001b[0m         \u001b[38;5;66;03m# can't write to a scalar, but indexing will be correct\u001b[39;00m\n\u001b[1;32m   4847\u001b[0m         result \u001b[38;5;241m=\u001b[39m arr[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/envs/default_env/lib/python3.10/site-packages/numpy/core/fromnumeric.py:2322\u001b[0m, in \u001b[0;36many\u001b[0;34m(a, axis, out, keepdims, where)\u001b[0m\n\u001b[1;32m   2317\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_any_dispatcher\u001b[39m(a, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, out\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, keepdims\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m,\n\u001b[1;32m   2318\u001b[0m                     where\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39m_NoValue):\n\u001b[1;32m   2319\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m (a, where, out)\n\u001b[0;32m-> 2322\u001b[0m \u001b[38;5;129m@array_function_dispatch\u001b[39m(_any_dispatcher)\n\u001b[1;32m   2323\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21many\u001b[39m(a, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, out\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, keepdims\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39m_NoValue, \u001b[38;5;241m*\u001b[39m, where\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39m_NoValue):\n\u001b[1;32m   2324\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   2325\u001b[0m \u001b[38;5;124;03m    Test whether any array element along a given axis evaluates to True.\u001b[39;00m\n\u001b[1;32m   2326\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   2410\u001b[0m \n\u001b[1;32m   2411\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m   2412\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _wrapreduction(a, np\u001b[38;5;241m.\u001b[39mlogical_or, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124many\u001b[39m\u001b[38;5;124m'\u001b[39m, axis, \u001b[38;5;28;01mNone\u001b[39;00m, out,\n\u001b[1;32m   2413\u001b[0m                           keepdims\u001b[38;5;241m=\u001b[39mkeepdims, where\u001b[38;5;241m=\u001b[39mwhere)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "updated_p_dict, hazard_ratios, f1_scores = find_threshold_vals(thresholds_by_f1_score)\n",
    "p_vals = pd.DataFrame.from_dict(updated_p_dict, orient = 'index')\n",
    "p_vals.columns = [\"P-value\"]\n",
    "hazard_ratios = pd.DataFrame.from_dict(hazard_ratios, orient = \"index\")\n",
    "hazard_ratios.columns = [\"Odds ratio\"]\n",
    "f1_scores = pd.DataFrame.from_dict(f1_scores, orient = 'index')\n",
    "f1_scores.columns = [\"F1 scores\"]\n",
    "thresholds = pd.DataFrame.from_dict(thresholds_by_f1_score, orient = 'index')\n",
    "thresholds.columns = [\"Thresholds\"]\n",
    "df = p_vals.join(hazard_ratios).join(thresholds).join(f1_scores)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
