{"cells":[{"cell_type":"code","execution_count":1,"id":"1aa0dd7c-f403-4246-80ad-525995615abd","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["# load packages\n","import pandas as pd\n","import numpy as np\n","from pandas.api.types import CategoricalDtype\n","import random\n","from random import shuffle"]},{"cell_type":"code","execution_count":null,"id":"40f30320-b693-4250-a6e1-44e77b3a6d8f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def load_variants(gene, dir_path = \"selected_genes/hcm/parsed_vep_files\"):\n","    \"\"\"\" loads csv with annotated variant information \"\"\" \n","    path = f'/mnt/project/{dir_path}/{gene}.csv'\n","    variants = pd.read_csv(path, sep = \",\")\n","\n","    variants = variants[[\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Name\", \"vep_consequence\", \"CADD_raw\", \"GERP++_RS\", \"phyloP100way_vertebrate\", 'trv', 'ds_ag', 'ds_al', 'ds_dg', 'ds_dl']]\n","\n","    return variants\n"]},{"cell_type":"code","execution_count":null,"id":"1dce2fa8-37ff-4aed-977b-3c1885754643","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def define_comp_predictors(gene, df, col):\n","    \"\"\" groups existing computation predictor scores (CADD_raw, phyloP100way_vertebrate, and GERP++_RS) by functional consequences \n","    and assigns them to variants based on this consequence. \"\"\"\n","\n","    col_means_grouped = df.groupby(by='vep_consequence')[col].mean()    \n","\n","\n","    for i in df.index:\n","        if pd.isnull(df[col][i]):\n","            if (pd.isnull(col_means_grouped[df['vep_consequence'][i]])) & (df['vep_consequence'][i] == 'synonymous_variant'):\n","               # what to do if no mean computational value exists for a Synonymous or Intronic consequence\n","                df.loc[[i],[col]] = comp_predictors_extreme_val(gene, df, col)\n","\n","            else:\n","                df.loc[[i],[col]] = col_means_grouped[df['vep_consequence'][i]]\n","                \n","    return df"]},{"cell_type":"code","execution_count":null,"id":"3a9ce24c-9025-4b0b-97e6-7dc4e7908f35","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def comp_predictors_extreme_val(gene, df, col):\n","    \"\"\" finds either lowest values or highest in a column for a computational score. returns this value \"\"\"\n","                    \n","    variants = df\n","        \n","    \n","    if col in ['CADD_raw', 'am_pathogenicity', 'REVEL']:\n","        val = min(variants[variants[col].notna()][col])\n","    elif (col == 'phyloP100way_vertebrate') | (col == 'GERP++_RS'):\n","        val = max(variants[variants[col].notna()][col])\n","    \n","    return val\n","                  \n","        "]},{"cell_type":"code","execution_count":null,"id":"066b4542-09a6-4f47-8c65-68047966d004","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def define_trv(gene,df,col):\n","    \"\"\" fills in trv values where they are missing, in accordance with their distribution in variants df\"\"\"\n","\n","    variants = load_variants(gene)\n","\n","\n","    \n","    distribution = variants[variants[col].notna()][col].tolist()\n","    empty_list = []\n","\n","    \n","    if (distribution == empty_list):\n","        df = df.drop([col], axis = 1)\n","        return df\n","        \n","\n","    for i in df.index:\n","        if pd.isnull(df[col][i]):\n","            df[col][i] = random.choice(distribution)\n","   \n","\n","    return df\n","                 "]},{"cell_type":"code","execution_count":null,"id":"bd1c5312-0400-40e0-91cc-f7ae64fcf7b2","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def add_gnomAD(parsed_gene):   \n","\n","    gnomAD_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/gnomAD_v4_variants_hcm.tsv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Filter\", \"AC\", \"AN\", \"AF\", \"faf_max\", \"empty\"])\n","    gene_with_gnomAD = parsed_gene.merge(gnomAD_csv, how='left', on = ['Chrom', 'Pos', 'Ref', 'Alt'])\n","    gene_with_gnomAD = gene_with_gnomAD[(gene_with_gnomAD['Filter'] == 'PASS')]\n","    gene_with_gnomAD = gene_with_gnomAD[['Chrom', 'Pos', 'Ref', 'Alt', \"Name\", \"vep_consequence\", \"CADD_raw\", \"GERP++_RS\", \"phyloP100way_vertebrate\", \"trv\", \"faf_max\", 'ds_ag', 'ds_al', 'ds_dg', 'ds_dl']]\n","\n","    return gene_with_gnomAD\n"]},{"cell_type":"code","execution_count":null,"id":"e8845372-efb0-40c4-a077-eec25ebc5379","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def add_AI(parsed_gene):\n","    AI_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/AlphaMissense_hg38.csv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"genome\", \"uniprot_id\", \"transcript_id\", \"protein_variant\", \"am_pathogenicity\", \"am_class\"])\n","    gene_with_AI = parsed_gene.merge(AI_csv, how='left', on = ['Chrom', 'Pos', 'Ref', 'Alt'])\n","    gene_with_AI = gene_with_AI[['Chrom', 'Pos', 'Ref', 'Alt', \"Name\", \"vep_consequence\", \"CADD_raw\", \"GERP++_RS\", \"phyloP100way_vertebrate\", \"trv\", \"faf_max\", \"am_pathogenicity\", 'ds_ag', 'ds_al', 'ds_dg', 'ds_dl']]\n","\n","    return gene_with_AI\n"]},{"cell_type":"code","execution_count":null,"id":"f425b211-45ba-42d7-a3dc-1de028db3c76","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_variants(gene):\n","    \"\"\" load and clean variants df and fill in missing values \"\"\"\n","    variants = load_variants(gene)\n","    variants = add_gnomAD(variants)\n","    variants = add_AI(variants)\n","    \n","\n","    \n","    # define deleterious mutations\n","    deleterious_list = [\"stop_gained\", \"start_lost\", \"splice_acceptor_variant\", \"splice_donor_variant\", \"splice_donor_region_variant\",\"splice_region_variant\",\"frameshift_variant\"]    \n","    deleterious = variants['vep_consequence'].apply(lambda x: any(substring in x for substring in deleterious_list))\n","    variants.loc[deleterious, 'vep_consequence'] = 'deleterious'\n","    \n","\n","    # filter for allele freq\n","    variants = variants.loc[(variants[\"faf_max\"] <= 0.002) | (variants[\"faf_max\"].isna())]\n","    \n","\n","    variants.loc[:, \"val\"] = 1\n","\n","    variants = define_comp_predictors(gene, variants, 'CADD_raw')\n","    variants = define_comp_predictors(gene, variants, 'phyloP100way_vertebrate')\n","    variants = define_comp_predictors(gene, variants, 'GERP++_RS') \n","    variants = define_comp_predictors(gene, variants, 'faf_max') \n","    variants = define_comp_predictors(gene, variants, 'am_pathogenicity')     \n","\n","\n","    return(variants)\n","\n","\n","\n","\n"]},{"cell_type":"code","execution_count":null,"id":"a51fd64e-d76a-4c8a-96fe-c31c0c2d97b6","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_carriers(gene, dir_path=\"selected_genes/hcm/carrier\"):\n","    \"\"\"retrieve variant-patient mapping file\"\"\"\n","    \n","    path = f'/mnt/project/{dir_path}/{gene}.ssv'\n","    carriers = pd.read_csv(path,\n","                       sep=\" \",\n","                       names=[\"Chrom\", \"empty\", \"Pos\", \"Ref\", \"Alt\", \"na\", \"Carriers\"],\n","                       usecols=lambda x: x not in ['na', 'empty'] )\n","    \n","    carriers.loc[:, \"Carriers\"] = carriers[\"Carriers\"].str.strip(\"|\")\n","    carriers.loc[:, \"Carriers\"]= carriers[\"Carriers\"].str.split(\"|\", expand = False)\n","    carriers = carriers.explode(\"Carriers\").rename(columns={\"Carriers\": \"Carrier\"})\n","    carriers = carriers[carriers['Carrier'].apply(lambda x: not x.startswith('W'))] # remove withdrawn\n","    \n","    \n","    return(carriers)\n","\n","\n","\n"]},{"cell_type":"code","execution_count":null,"id":"f7908657-3f95-4c61-9790-1f09d274a713","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_joined(gene, **paths):\n","    \"\"\"retrieves both variant and mapping files and joins them together by variant\"\"\"\n","    \n","    variants = get_variants(gene, paths[\"var_path\"]) if \"var_path\" in paths else get_variants(gene)\n","    carriers = get_carriers(gene, paths[\"car_path\"]) if \"car_path\" in paths else get_carriers(gene)\n","    \n","  \n","    \n","    variants = variants.astype({'Chrom': 'object', 'Pos':'object'})\n","    \n","    carriers = carriers.astype({'Pos':'object'})\n","    \n","    ann_carriers = carriers.merge(variants, on=[\"Chrom\", \"Pos\", \"Ref\", \"Alt\"], how = \"inner\")  \n","    \n","    \n","    return(ann_carriers)\n"]},{"cell_type":"code","execution_count":null,"id":"a39b398b-c4ed-41ff-bb45-d874d1fd624a","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_most_severe(ann_carriers):\n","    \n","    \"\"\"accepts a filtered patient-variant df.\n","    selects and retains only the most severe variant per patient. returns this df.\"\"\"\n","    \n"," \n","    \n","    # order variant consequence by general severity (deleterious > missense > synonymous)\n","    consequence_cats = CategoricalDtype(categories=[\"synonymous_variant\", \"missense_variant\", \"deleterious\"], ordered=True)\n","    ann_carriers.loc[:, \"vep_consequence\"] = ann_carriers[\"vep_consequence\"].astype(consequence_cats)\n","    \n","    # choose maximum AF for each variant, where possible\n","    ann_carriers.loc[:, \"max_AF\"] = ann_carriers.loc[:, [\"faf_max\"]].max(axis=1, skipna=True)\n","     \n","    # create a random tie breaker index\n","    shuffled_index = list(range(0,len(ann_carriers)))\n","    shuffle(shuffled_index)\n","    ann_carriers.loc[:, \"tie_breaker\"] = shuffled_index\n","    \n","    # group by eid, sort by consequence > CADD > confidence (for del. vars) > AF > tiebreaker\n","    # select top variant for each group\n","    ann_carriers_grouped = ann_carriers.groupby(\"Carrier\", group_keys=False).apply(pd.DataFrame.sort_values, [\"vep_consequence\", \"CADD_raw\", \"max_AF\", \"tie_breaker\"], ascending=[False, False, True, True])\n","    ann_carriers_grouped = ann_carriers_grouped.drop_duplicates(\"Carrier\", keep=\"first\")\n","    \n","    # important!! return categorical variables to normal objects so they don't clog pivot memory\n","    ann_carriers_grouped.loc[:, \"vep_consequence\"] = ann_carriers_grouped[\"vep_consequence\"].astype(object)\n","    \n","\n","\n","    \n","    return(ann_carriers_grouped)"]},{"cell_type":"code","execution_count":null,"id":"1d7a4dc5-0057-4c0c-aeac-a3dfae2dcf3b","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def reshape_consequence(ann_carriers, rename_id=True):\n","    \"\"\"\n","    accepts filtered and prepped patient-variant df\n","    returns a reshaped df that separates consequences into unique columns\n","\n","    \"\"\"\n","    ann_carriers = ann_carriers.reset_index(drop = True)\n","    \n","    for col in ['CADD_raw', 'phyloP100way_vertebrate',  \"GERP++_RS\", \"trv\",'ds_ag', 'ds_al', 'ds_dg', 'ds_dl', \"am_pathogenicity\"]:\n","        ann_carriers[col] = ann_carriers[col].fillna(\"temp\")\n","        \n","    ann_carriers = ann_carriers[[\"Carrier\", \"Name\", \"CADD_raw\", \"phyloP100way_vertebrate\", \"GERP++_RS\", \"trv\", 'ds_ag', 'ds_al', 'ds_dg', 'ds_dl', \"am_pathogenicity\", \"vep_consequence\", \"val\"]]\n","\n","    \n","    \n","    ann_carriers = ann_carriers.pivot_table(index=[\"Carrier\", \"Name\", \"CADD_raw\", \"phyloP100way_vertebrate\", \"GERP++_RS\", \"trv\", 'ds_ag', 'ds_al', 'ds_dg', 'ds_dl', \"am_pathogenicity\"], values=\"val\", columns=\"vep_consequence\", fill_value=0)\n","    ann_carriers.to_csv(\"ann_carriers_end.csv\")\n","\n","    \n","    new_names={\"Name\": \"variant_id\", \"Carrier\":\"eid\"}\n","    \n","    ann_carriers = ann_carriers.reset_index()\n","\n","    for col in ['CADD_raw', 'phyloP100way_vertebrate',  \"GERP++_RS\", \"trv\", 'ds_ag', 'ds_al', 'ds_dg', 'ds_dl', \"am_pathogenicity\"]:\n","        ann_carriers[col] = ann_carriers[col].replace('temp', None)\n","\n","\n","\n","\n","    # this variable is set to true whenever get_most_severe was run before this function\n","    if rename_id:\n","        new_names[\"Name\"] = \"most_severe_variant\"\n","    \n","\n","\n","\n","    \n","    return(ann_carriers)\n","\n","\n"]},{"cell_type":"code","execution_count":null,"id":"0ffec0a3-841f-443c-813a-2430a48021d8","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def patient_var_mappings(gene, most_severe=True, **paths):\n","\n","    \"\"\"\n","    accepts gene name and some combination of var_path and car_path directories.\n","    returns a pandas df with one row per patient detailing their most severe variant\n","    \"\"\"\n","    annotated_carriers = get_joined(gene, **paths)\n","    \n","    if most_severe:\n","        annotated_carriers = get_most_severe(annotated_carriers)\n","        \n","  \n","    return(reshape_consequence(annotated_carriers, rename_id=most_severe))\n","\n"]},{"cell_type":"code","execution_count":null,"id":"d2a0ef3a-6929-40ae-91e8-3ca3b1e198c3","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def merge_all_characteristics(gene):\n","    \n","    # read in and clean phenotype, principal components, and polygenic risk score dataframes\n","    project_path = f'/mnt/project/selected_genes/hcm/csv_files/'\n","    prs_csv = pd.read_csv(project_path+'applied_hcm_prs.csv')\n","    principal_components_csv = pd.read_csv(project_path+'principal_components.csv')\n","    phenotypic_info_csv = pd.read_csv(project_path+'hypertrophic_df.csv')\n","\n","    prs_csv['Carrier'] = prs_csv['eid'].astype(str)\n","    prs_csv = prs_csv.drop(['eid'], axis=1)\n","    principal_components_csv['Carrier'] = principal_components_csv['eid'].astype(str)\n","    principal_components_csv = principal_components_csv.drop(['eid'], axis=1)\n","\n","    phenotypic_info_csv['Carrier'] = phenotypic_info_csv['eid'].astype(str)  \n","    phenotypic_info_csv = phenotypic_info_csv.drop(['eid'], axis=1)\n","    \n","    \n","    g = patient_var_mappings(gene).reset_index()\n","\n","\n","    # join all dataframes\n","    prs_joined = prs_csv.set_index('Carrier').join(g.set_index('Carrier'), how = 'left')\n","    principal_prs_joined = principal_components_csv.set_index('Carrier').join(prs_joined, how = 'outer')\n","    \n","    all_joined = phenotypic_info_csv.set_index('Carrier').join(principal_prs_joined, how = 'outer')\n","    \n","\n","    # fill in missing values for patients without annotated variants\n","    if 'trv' in all_joined.columns:\n","        all_joined = define_trv(gene,all_joined,\"trv\")\n","    all_joined[\"synonymous_variant\"] = all_joined[\"synonymous_variant\"].fillna(\"0\")\n","    all_joined[\"missense_variant\"] = all_joined[\"missense_variant\"].fillna(\"0\")\n","    all_joined[\"deleterious\"] = all_joined[\"deleterious\"].fillna(\"0\")\n","    all_joined[\"ds_ag\"] = all_joined[\"ds_ag\"].fillna(\"0\")\n","    all_joined[\"ds_al\"] = all_joined[\"ds_al\"].fillna(\"0\")\n","    all_joined[\"ds_dg\"] = all_joined[\"ds_dg\"].fillna(\"0\")\n","    all_joined[\"ds_dl\"] = all_joined[\"ds_dl\"].fillna(\"0\")\n","    all_joined[\"CADD_raw\"] = all_joined[\"CADD_raw\"].fillna(comp_predictors_extreme_val(gene, all_joined, \"CADD_raw\"))\n","    all_joined[\"GERP++_RS\"] = all_joined[\"GERP++_RS\"].fillna(comp_predictors_extreme_val(gene, all_joined, \"GERP++_RS\"))\n","    all_joined[\"phyloP100way_vertebrate\"] = all_joined[\"phyloP100way_vertebrate\"].fillna(comp_predictors_extreme_val(gene, all_joined, \"phyloP100way_vertebrate\"))\n","    all_joined[\"am_pathogenicity\"] = all_joined[\"am_pathogenicity\"].fillna(comp_predictors_extreme_val(gene, all_joined, \"am_pathogenicity\"))\n","\n","    all_joined = all_joined.drop(['Unnamed: 0', 'birth_date', 'death_age'], axis = 1)\n","    all_joined[\"is_family_hist\"] = np.where(all_joined[\"is_family_hist\"]==True, 1, 0)\n","    all_joined[\"is_hcm\"] = np.where(all_joined[\"is_hcm\"]==True, 1, 0)\n","\n","\n","    \n","    # hacks\n","    all_joined = all_joined[all_joined['duration'] > 20]\n","\n","\n","    return all_joined \n","    "]},{"cell_type":"code","execution_count":61,"id":"fec9ff5f-27d3-4e6a-a30f-0d773ef0e580","metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stderr","output_type":"stream","text":["/tmp/ipykernel_411/1223142907.py:3: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n","  gnomAD_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/gnomAD_v4_variants_hcm.tsv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Filter\", \"AC\", \"AN\", \"AF\", \"faf_max\", \"empty\"])\n","/tmp/ipykernel_411/352821128.py:2: DtypeWarning: Columns (1,8) have mixed types. Specify dtype option on import or set low_memory=False.\n","  AI_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/AlphaMissense_hg38.csv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"genome\", \"uniprot_id\", \"transcript_id\", \"protein_variant\", \"am_pathogenicity\", \"am_class\"])\n","/tmp/ipykernel_411/386404196.py:10: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`\n","  ann_carriers.loc[:, \"vep_consequence\"] = ann_carriers[\"vep_consequence\"].astype(consequence_cats)\n","/tmp/ipykernel_411/386404196.py:26: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`\n","  ann_carriers_grouped.loc[:, \"vep_consequence\"] = ann_carriers_grouped[\"vep_consequence\"].astype(object)\n","/tmp/ipykernel_411/586892702.py:19: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df[col][i] = random.choice(distribution)\n"]},{"ename":"KeyError","evalue":"'synonymous_variant'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)","File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n","File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n","File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n","File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n","File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n","\u001b[0;31mKeyError\u001b[0m: 'synonymous_variant'","\nThe above exception was the direct cause of the following exception:\n","\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)","Cell \u001b[0;32mIn[61], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m g \u001b[38;5;129;01min\u001b[39;00m genes:\n\u001b[1;32m      4\u001b[0m     file_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mg\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_gnomad_AI_data.csv\u001b[39m\u001b[38;5;124m'\u001b[39m  \n\u001b[0;32m----> 5\u001b[0m     all_carriers \u001b[38;5;241m=\u001b[39m \u001b[43mmerge_all_characteristics\u001b[49m\u001b[43m(\u001b[49m\u001b[43mg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m     all_carriers\u001b[38;5;241m.\u001b[39mto_csv(file_name, index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n","Cell \u001b[0;32mIn[29], line 31\u001b[0m, in \u001b[0;36mmerge_all_characteristics\u001b[0;34m(gene)\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtrv\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m all_joined\u001b[38;5;241m.\u001b[39mcolumns:\n\u001b[1;32m     30\u001b[0m     all_joined \u001b[38;5;241m=\u001b[39m define_trv(gene,all_joined,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtrv\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 31\u001b[0m all_joined[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msynonymous_variant\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mall_joined\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msynonymous_variant\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mfillna(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m0\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     32\u001b[0m all_joined[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmissense_variant\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m all_joined[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmissense_variant\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mfillna(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m0\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     33\u001b[0m all_joined[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdeleterious\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m all_joined[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdeleterious\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mfillna(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m0\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n","File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/frame.py:3807\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3805\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3807\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3808\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3809\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n","File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/indexes/base.py:3804\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3804\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3808\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3809\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n","\u001b[0;31mKeyError\u001b[0m: 'synonymous_variant'"]}],"source":["genes = [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\"]\n","\n","for g in genes:\n","    file_name = f'{g}_gnomad_AI_data.csv'  \n","    all_carriers = merge_all_characteristics(g)\n","   \n","    all_carriers.to_csv(file_name, index=False)\n","\n","    "]},{"cell_type":"code","execution_count":null,"id":"b4b06a18-2f59-490f-9acc-292d07357b0f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["!dx mkdir -p \"Cassa Lab Shared Project:selected_genes/hcm/lifelines_data/\"\n","!dx cd /selected_genes/hcm/lifelines_data/\n","!dx upload *.csv"]},{"cell_type":"code","execution_count":null,"id":"98e2d920-6365-45cc-a9d3-f0f18600a092","metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.13"}},"nbformat":4,"nbformat_minor":5}
