{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.9.16"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "code", "source": "# load packages\nimport pandas as pd\nimport numpy as np\nfrom pandas.api.types import CategoricalDtype\nimport random\nfrom random import shuffle", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "1aa0dd7c-f403-4246-80ad-525995615abd"}, {"cell_type": "code", "source": "def load_variants(gene, dir_path = \"selected_genes/hcm/parsed_vep_files\"):\n    \"\"\"\" loads csv with annotated variant information \"\"\" \n    path = f'/mnt/project/{dir_path}/{gene}.csv'\n    variants = pd.read_csv(path, sep = \",\")\n    \n    \n    variants = variants[[\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Name\", 'vep_consequence', 'GERP++_RS', 'trv', 'CADD_raw', 'phyloP100way_vertebrate']]\n   \n    deleterious_list = [\"stop_gained\", \"start_lost\", \"splice_acceptor_variant\", \"splice_donor_variant\", \"splice_donor_region_variant\",\"splice_region_variant\",\"frameshift_variant\"]    \n    deleterious = variants['vep_consequence'].apply(lambda x: any(substring in x for substring in deleterious_list))\n    variants.loc[deleterious, 'vep_consequence'] = 'deleterious'\n    \n    variants = variants[(variants['vep_consequence'] == 'deleterious') | (variants['vep_consequence'] == 'missense_variant')]\n\n    return variants\n\n\n\n\n\ngenes = [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PTPN11\", \"TNNI3\", \"TTR\", \"PLN\"]\nfor g in genes:\n    load_variants(g).to_csv(g+\".csv\")\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 30, "outputs": [], "id": "40f30320-b693-4250-a6e1-44e77b3a6d8f"}, {"cell_type": "code", "source": "def add_gnomAD(parsed_gene):   \n\n    gnomAD_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/gnomAD_v4_variants_hcm.tsv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Filter\", \"AC\", \"AN\", \"AF\", \"faf_max\", \"empty\"])\n    gnomAD_csv = gnomAD_csv[[\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Filter\", \"faf_max\",]]\n    gene_with_gnomAD = parsed_gene.merge(gnomAD_csv, how='left', on = ['Chrom', 'Pos', 'Ref', 'Alt'])\n    gene_with_gnomAD = gene_with_gnomAD[(gene_with_gnomAD['Filter'] == 'PASS')]\n    gene_with_gnomAD = gene_with_gnomAD[['Chrom', 'Pos', 'Ref', 'Alt', \"Name\", \"vep_consequence\", \"faf_max\", 'GERP++_RS', 'trv', 'CADD_raw', 'phyloP100way_vertebrate']]\n    return gene_with_gnomAD\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "bd1c5312-0400-40e0-91cc-f7ae64fcf7b2"}, {"cell_type": "code", "source": "def add_AI(parsed_gene):\n    AI_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/AlphaMissense_hg38.csv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"genome\", \"uniprot_id\", \"transcript_id\", \"protein_variant\", \"am_pathogenicity\", \"am_class\"])\n    gene_with_AI = parsed_gene.merge(AI_csv, how='left', on = ['Chrom', 'Pos', 'Ref', 'Alt'])\n    gene_with_AI = gene_with_AI[['Chrom', 'Pos', 'Ref', 'Alt', \"Name\", \"vep_consequence\", \"faf_max\", \"am_pathogenicity\", 'GERP++_RS', 'trv', 'CADD_raw', 'phyloP100way_vertebrate']]\n\n    return gene_with_AI\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "e8845372-efb0-40c4-a077-eec25ebc5379"}, {"cell_type": "code", "source": "def get_variants(gene):\n    \"\"\" load and clean variants df and fill in missing values \"\"\"\n    variants = load_variants(gene)\n    variants = add_gnomAD(variants)\n    variants = add_AI(variants)\n    \n    \n    \n\n    \n    # define deleterious mutations\n\n\n    # filter for allele freq\n    variants = variants.loc[(variants[\"faf_max\"] <= 0.002) | (variants[\"faf_max\"].isna())]\n    \n\n    variants.loc[:, \"val\"] = 1\n\n    return(variants)\n\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "f425b211-45ba-42d7-a3dc-1de028db3c76"}, {"cell_type": "code", "source": "def get_carriers(gene, dir_path=\"selected_genes/hcm/carrier\"):\n    \"\"\"retrieve variant-patient mapping file\"\"\"\n    \n    path = f'/mnt/project/{dir_path}/{gene}.ssv'\n    carriers = pd.read_csv(path,\n                       sep=\" \",\n                       names=[\"Chrom\", \"empty\", \"Pos\", \"Ref\", \"Alt\", \"na\", \"Carriers\"],\n                       usecols=lambda x: x not in ['na', 'empty'] )\n    \n    carriers.loc[:, \"Carriers\"] = carriers[\"Carriers\"].str.strip(\"|\")\n    carriers.loc[:, \"Carriers\"]= carriers[\"Carriers\"].str.split(\"|\", expand = False)\n    carriers = carriers.explode(\"Carriers\").rename(columns={\"Carriers\": \"Carrier\"})\n    carriers = carriers[carriers['Carrier'].apply(lambda x: not x.startswith('W'))] # remove withdrawn\n    \n    \n    return(carriers)", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "a51fd64e-d76a-4c8a-96fe-c31c0c2d97b6"}, {"cell_type": "code", "source": "def get_joined(gene, **paths):\n    \"\"\"retrieves both variant and mapping files and joins them together by variant\"\"\"\n    \n    variants = get_variants(gene, paths[\"var_path\"]) if \"var_path\" in paths else get_variants(gene)\n    carriers = get_carriers(gene, paths[\"car_path\"]) if \"car_path\" in paths else get_carriers(gene)\n  \n    \n    variants = variants.astype({'Chrom': 'object', 'Pos':'object'})\n    \n    carriers = carriers.astype({'Pos':'object'})\n    \n    ann_carriers = carriers.merge(variants, on=[\"Chrom\", \"Pos\", \"Ref\", \"Alt\"], how = \"inner\")  \n    \n    \n    return(ann_carriers)\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "f7908657-3f95-4c61-9790-1f09d274a713"}, {"cell_type": "code", "source": "def get_most_severe(ann_carriers):\n    \n    \"\"\"accepts a filtered patient-variant df.\n    selects and retains only the most severe variant per patient. returns this df.\"\"\"\n    \n \n    \n    # order variant consequence by general severity (deleterious > missense)\n    consequence_cats = CategoricalDtype(categories=[\"missense_variant\", \"deleterious\"], ordered=True)\n    ann_carriers.loc[:, \"vep_consequence\"] = ann_carriers[\"vep_consequence\"].astype(consequence_cats)\n    \n    # choose maximum AF for each variant, where possible\n    ann_carriers.loc[:, \"max_AF\"] = ann_carriers.loc[:, [\"faf_max\"]].max(axis=1, skipna=True)\n     \n    # create a random tie breaker index\n    shuffled_index = list(range(0,len(ann_carriers)))\n    shuffle(shuffled_index)\n    ann_carriers.loc[:, \"tie_breaker\"] = shuffled_index\n    \n    # group by eid, sort by consequence > CADD > confidence (for del. vars) > AF > tiebreaker\n    # select top variant for each group\n    ann_carriers_grouped = ann_carriers.groupby(\"Carrier\", group_keys=False).apply(pd.DataFrame.sort_values, [\"vep_consequence\", \"max_AF\", \"tie_breaker\"], ascending=[False, True, True])\n    ann_carriers_grouped = ann_carriers_grouped.drop_duplicates(\"Carrier\", keep=\"first\")\n    \n    # important!! return categorical variables to normal objects so they don't clog pivot memory\n    ann_carriers_grouped.loc[:, \"vep_consequence\"] = ann_carriers_grouped[\"vep_consequence\"].astype(object)\n    \n\n\n    \n    return(ann_carriers_grouped)", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "a39b398b-c4ed-41ff-bb45-d874d1fd624a"}, {"cell_type": "code", "source": "def reshape_consequence(ann_carriers, rename_id=True):\n    \"\"\"\n    accepts filtered and prepped patient-variant df\n    returns a reshaped df that separates consequences into unique columns\n\n    \"\"\"\n    ann_carriers = ann_carriers.reset_index(drop = True)\n    \n    for col in [\"am_pathogenicity\"]:\n        ann_carriers[col] = ann_carriers[col].fillna(\"temp\")\n        \n    ann_carriers = ann_carriers[[\"Carrier\", \"Name\", \"am_pathogenicity\", \"vep_consequence\", \"val\", 'GERP++_RS', 'trv', 'CADD_raw', 'phyloP100way_vertebrate']]\n    ann_carriers = ann_carriers.pivot_table(index=[\"Carrier\", \"Name\", \"am_pathogenicity\", 'GERP++_RS', 'trv', 'CADD_raw', 'phyloP100way_vertebrate'], values=\"val\", columns=\"vep_consequence\", fill_value=0)\n    new_names={\"Name\": \"variant_id\", \"Carrier\":\"eid\"}\n\n        \n    # this variable is set to true whenever get_most_severe was run before this function\n    if rename_id:\n        new_names[\"Name\"] = \"most_severe_variant\"\n        \n        \n    \n\n\n\n    \n    return(ann_carriers)\n\n\n\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "1d7a4dc5-0057-4c0c-aeac-a3dfae2dcf3b"}, {"cell_type": "code", "source": "def patient_var_mappings(gene, most_severe=True, **paths):\n\n    \"\"\"\n    accepts gene name and some combination of var_path and car_path directories.\n    returns a pandas df with one row per patient detailing their most severe variant\n    \"\"\"\n    annotated_carriers = get_joined(gene, **paths)\n    \n    if most_severe:\n        annotated_carriers = get_most_severe(annotated_carriers)\n        \n  \n    return(reshape_consequence(annotated_carriers, rename_id=most_severe))\n\n\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "0ffec0a3-841f-443c-813a-2430a48021d8"}, {"cell_type": "code", "source": "def merge_all_characteristics(gene):\n    \n    # read in and clean phenotype, principal components, and polygenic risk score dataframes\n    project_path = f'/mnt/project/selected_genes/hcm/csv_files/'\n    prs_csv = pd.read_csv(project_path+'applied_hcm_prs.csv')\n    principal_components_csv = pd.read_csv(project_path+'principal_components.csv')\n    phenotypic_info_csv = pd.read_csv(project_path+'hypertrophic_df.csv')\n\n    prs_csv['Carrier'] = prs_csv['eid'].astype(str)\n    prs_csv = prs_csv.drop(['eid'], axis=1)\n    \n    \n    \n    \n    \n    principal_components_csv['Carrier'] = principal_components_csv['eid'].astype(str)\n    principal_components_csv = principal_components_csv.drop(['eid',  'principal_component_2','principal_component_3', 'principal_component_5', 'principal_component_6', 'principal_component_7', 'principal_component_8', 'principal_component_9', 'principal_component_10'], axis=1)\n\n    phenotypic_info_csv['Carrier'] = phenotypic_info_csv['eid'].astype(str)  \n    phenotypic_info_csv = phenotypic_info_csv.drop(['eid'], axis=1)\n    \n    \n    g = patient_var_mappings(gene)\n    g = g.reset_index()\n    prs_csv = prs_csv.reset_index()\n\n\n\n    # join all dataframes\n    prs_joined = prs_csv.merge(g, how = 'left')\n    principal_prs_joined = principal_components_csv.merge(prs_joined, how = 'right')\n    \n    all_joined = phenotypic_info_csv.merge(principal_prs_joined, how = 'right')\n    \n\n    # fill in missing values for patients without annotated variants\n    all_joined = all_joined[(all_joined['deleterious'] == True) | (all_joined['missense_variant'] == True)]\n\n    all_joined[\"missense_variant\"] = all_joined[\"missense_variant\"].fillna(\"0\")\n    all_joined[\"deleterious\"] = all_joined[\"deleterious\"].fillna(\"0\")\n\n    all_joined[\"is_family_hist\"] = np.where(all_joined[\"is_family_hist\"]==True, 1, 0)\n    all_joined[\"is_hcm\"] = np.where(all_joined[\"is_hcm\"]==True, 1, 0)\n    all_joined = all_joined.drop([\"death_age\",\"Unnamed: 0\", \"birth_date\", 'deleterious', 'missense_variant'], axis = 1)\n\n\n    return all_joined \n    ", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "d2a0ef3a-6929-40ae-91e8-3ca3b1e198c3"}, {"cell_type": "code", "source": "genes = [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PTPN11\", \"TNNI3\", \"TTR\", \"PLN\"]\nfor g in genes:\n    file_name = f'{g}.csv'  \n    all_carriers = merge_all_characteristics(g)\n   \n    all_carriers.to_csv(file_name, index=False)\n\n    ", "metadata": {"trusted": true, "tags": []}, "execution_count": 29, "outputs": [{"name": "stderr", "text": "/tmp/ipykernel_171/49528828.py:3: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n  gnomAD_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/gnomAD_v4_variants_hcm.tsv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Filter\", \"AC\", \"AN\", \"AF\", \"faf_max\", \"empty\"])\n/tmp/ipykernel_171/3262844094.py:2: DtypeWarning: Columns (1,8) have mixed types. Specify dtype option on import or set low_memory=False.\n  AI_csv = pd.read_csv(f'/mnt/project/selected_genes/hcm/csv_files/AlphaMissense_hg38.csv', sep = \"\\t\", names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"genome\", \"uniprot_id\", \"transcript_id\", \"protein_variant\", \"am_pathogenicity\", \"am_class\"])\n", "output_type": "stream"}, {"traceback": ["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m", "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)", "Cell \u001b[0;32mIn[29], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m g \u001b[38;5;129;01min\u001b[39;00m genes:\n\u001b[1;32m      3\u001b[0m     file_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mg\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m'\u001b[39m  \n\u001b[0;32m----> 4\u001b[0m     all_carriers \u001b[38;5;241m=\u001b[39m \u001b[43mmerge_all_characteristics\u001b[49m\u001b[43m(\u001b[49m\u001b[43mg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      6\u001b[0m     all_carriers\u001b[38;5;241m.\u001b[39mto_csv(file_name, index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n", "Cell \u001b[0;32mIn[24], line 23\u001b[0m, in \u001b[0;36mmerge_all_characteristics\u001b[0;34m(gene)\u001b[0m\n\u001b[1;32m     19\u001b[0m phenotypic_info_csv[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCarrier\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m phenotypic_info_csv[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124meid\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mstr\u001b[39m)  \n\u001b[1;32m     20\u001b[0m phenotypic_info_csv \u001b[38;5;241m=\u001b[39m phenotypic_info_csv\u001b[38;5;241m.\u001b[39mdrop([\u001b[38;5;124m'\u001b[39m\u001b[38;5;124meid\u001b[39m\u001b[38;5;124m'\u001b[39m], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[0;32m---> 23\u001b[0m g \u001b[38;5;241m=\u001b[39m \u001b[43mpatient_var_mappings\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     24\u001b[0m g \u001b[38;5;241m=\u001b[39m g\u001b[38;5;241m.\u001b[39mreset_index()\n\u001b[1;32m     25\u001b[0m prs_csv \u001b[38;5;241m=\u001b[39m prs_csv\u001b[38;5;241m.\u001b[39mreset_index()\n", "Cell \u001b[0;32mIn[10], line 7\u001b[0m, in \u001b[0;36mpatient_var_mappings\u001b[0;34m(gene, most_severe, **paths)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpatient_var_mappings\u001b[39m(gene, most_severe\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mpaths):\n\u001b[1;32m      3\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;124;03m    accepts gene name and some combination of var_path and car_path directories.\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;124;03m    returns a pandas df with one row per patient detailing their most severe variant\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m----> 7\u001b[0m     annotated_carriers \u001b[38;5;241m=\u001b[39m \u001b[43mget_joined\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mpaths\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m most_severe:\n\u001b[1;32m     10\u001b[0m         annotated_carriers \u001b[38;5;241m=\u001b[39m get_most_severe(annotated_carriers)\n", "Cell \u001b[0;32mIn[7], line 4\u001b[0m, in \u001b[0;36mget_joined\u001b[0;34m(gene, **paths)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_joined\u001b[39m(gene, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mpaths):\n\u001b[1;32m      2\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"retrieves both variant and mapping files and joins them together by variant\"\"\"\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m     variants \u001b[38;5;241m=\u001b[39m get_variants(gene, paths[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar_path\u001b[39m\u001b[38;5;124m\"\u001b[39m]) \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar_path\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m paths \u001b[38;5;28;01melse\u001b[39;00m \u001b[43mget_variants\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m     carriers \u001b[38;5;241m=\u001b[39m get_carriers(gene, paths[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcar_path\u001b[39m\u001b[38;5;124m\"\u001b[39m]) \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcar_path\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m paths \u001b[38;5;28;01melse\u001b[39;00m get_carriers(gene)\n\u001b[1;32m      8\u001b[0m     variants \u001b[38;5;241m=\u001b[39m variants\u001b[38;5;241m.\u001b[39mastype({\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mChrom\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mobject\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPos\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mobject\u001b[39m\u001b[38;5;124m'\u001b[39m})\n", "Cell \u001b[0;32mIn[5], line 18\u001b[0m, in \u001b[0;36mget_variants\u001b[0;34m(gene)\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m# define deleterious mutations\u001b[39;00m\n\u001b[1;32m     12\u001b[0m \n\u001b[1;32m     13\u001b[0m \n\u001b[1;32m     14\u001b[0m \u001b[38;5;66;03m# filter for allele freq\u001b[39;00m\n\u001b[1;32m     15\u001b[0m variants \u001b[38;5;241m=\u001b[39m variants\u001b[38;5;241m.\u001b[39mloc[(variants[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfaf_max\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.002\u001b[39m) \u001b[38;5;241m|\u001b[39m (variants[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfaf_max\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39misna())]\n\u001b[0;32m---> 18\u001b[0m \u001b[43mvariants\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mval\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m(variants)\n", "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:818\u001b[0m, in \u001b[0;36m_LocationIndexer.__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m    815\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_has_valid_setitem_indexer(key)\n\u001b[1;32m    817\u001b[0m iloc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miloc\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39miloc\n\u001b[0;32m--> 818\u001b[0m \u001b[43miloc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_setitem_with_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\n", "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:1718\u001b[0m, in \u001b[0;36m_iLocIndexer._setitem_with_indexer\u001b[0;34m(self, indexer, value, name)\u001b[0m\n\u001b[1;32m   1716\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj):\n\u001b[1;32m   1717\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_list_like_indexer(value):\n\u001b[0;32m-> 1718\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1719\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcannot set a frame with no \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1720\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdefined index and a scalar\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1721\u001b[0m         )\n\u001b[1;32m   1722\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj[key] \u001b[38;5;241m=\u001b[39m value\n\u001b[1;32m   1723\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n", "\u001b[0;31mValueError\u001b[0m: cannot set a frame with no defined index and a scalar"], "ename": "ValueError", "evalue": "cannot set a frame with no defined index and a scalar", "output_type": "error"}], "id": "fec9ff5f-27d3-4e6a-a30f-0d773ef0e580"}, {"cell_type": "code", "source": "!dx mkdir -p \"Cassa Lab Shared Project:selected_genes/hcm/lifelines_data/\"\n!dx cd /selected_genes/hcm/lifelines_data/\n!dx upload *.csv", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "b4b06a18-2f59-490f-9acc-292d07357b0f"}, {"cell_type": "code", "source": "MYBPC3", "metadata": {"trusted": true, "tags": []}, "execution_count": 23, "outputs": [{"execution_count": 23, "output_type": "execute_result", "data": {"text/plain": "        Unnamed: 0  sex  is_family_hist  is_hcm   duration        age  \\\n300          300.0  0.0               0       0  79.950719  79.950719   \n356          356.0  1.0               0       0  78.000000  78.000000   \n402          402.0  0.0               0       0  74.863792  80.117728   \n572          572.0  0.0               0       0  67.370294  67.370294   \n726          726.0  0.0               1       0  78.450376  78.450376   \n...            ...  ...             ...     ...        ...        ...   \n502079    502079.0  0.0               0       0  77.864476  77.864476   \n502084    502084.0  0.0               0       0  61.779603  61.779603   \n502222    502222.0  1.0               0       0  61.535934  61.535934   \n502299    502299.0  0.0               1       0  72.703628  72.703628   \n502342    502342.0  1.0               0       0  77.032170  77.032170   \n\n        Carrier  principal_component_1  principal_component_4     index  \\\n300     1014971                2.25973              -44.25070    1443.0   \n356     1017630               57.43850               10.55800    1697.0   \n402     1020263              383.80500                5.21918    1952.0   \n572     1028911              -13.13840                1.26250    2791.0   \n726     1036853               -5.87859               -4.38288    3548.0   \n...         ...                    ...                    ...       ...   \n502079  6011707              -13.30810                2.01273  482436.0   \n502084  6011959              -11.68130               -2.81797  482460.0   \n502222  6018974              -14.20980                2.78787  483140.0   \n502299  6022063              -12.70350                2.92162  483439.0   \n502342  6023765              -15.71110                1.60306  483600.0   \n\n        prs_score             Name am_pathogenicity  GERP++_RS  trv  CADD_raw  \\\n300      0.000658  11-47341995-C-T           0.6363       5.13  0.0  4.076783   \n356     -0.013064  11-47338586-C-T           0.0882       3.54  0.0  0.968143   \n402      0.020625  11-47350068-C-T           0.1707       3.43  0.0  1.672304   \n572     -0.012814  11-47332912-A-G           0.3284       5.25  0.0  3.094806   \n726     -0.012519  11-47342016-G-A           0.6316       5.13  0.0  4.282963   \n...           ...              ...              ...        ...  ...       ...   \n502079   0.015772  11-47342734-C-T           0.5758       4.63  0.0  3.708413   \n502084  -0.027074  11-47333208-C-T           0.4444       5.36  0.0  3.801036   \n502222  -0.000138  11-47338579-G-A           0.1536       1.05  0.0  1.961707   \n502299   0.006107  11-47332912-A-G           0.3284       5.25  0.0  3.094806   \n502342  -0.003158  11-47332912-A-G           0.3284       5.25  0.0  3.094806   \n\n        phyloP100way_vertebrate  deleterious  missense_variant  \n300                       7.531          0.0               1.0  \n356                       1.537          0.0               1.0  \n402                       0.511          0.0               1.0  \n572                       7.144          0.0               1.0  \n726                       2.437          0.0               1.0  \n...                         ...          ...               ...  \n502079                    7.506          0.0               1.0  \n502084                    7.556          0.0               1.0  \n502222                    0.428          0.0               1.0  \n502299                    7.144          0.0               1.0  \n502342                    7.144          0.0               1.0  \n\n[8044 rows x 19 columns]", "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Unnamed: 0</th>\n      <th>sex</th>\n      <th>is_family_hist</th>\n      <th>is_hcm</th>\n      <th>duration</th>\n      <th>age</th>\n      <th>Carrier</th>\n      <th>principal_component_1</th>\n      <th>principal_component_4</th>\n      <th>index</th>\n      <th>prs_score</th>\n      <th>Name</th>\n      <th>am_pathogenicity</th>\n      <th>GERP++_RS</th>\n      <th>trv</th>\n      <th>CADD_raw</th>\n      <th>phyloP100way_vertebrate</th>\n      <th>deleterious</th>\n      <th>missense_variant</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>300</th>\n      <td>300.0</td>\n      <td>0.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>79.950719</td>\n      <td>79.950719</td>\n      <td>1014971</td>\n      <td>2.25973</td>\n      <td>-44.25070</td>\n      <td>1443.0</td>\n      <td>0.000658</td>\n      <td>11-47341995-C-T</td>\n      <td>0.6363</td>\n      <td>5.13</td>\n      <td>0.0</td>\n      <td>4.076783</td>\n      <td>7.531</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>356</th>\n      <td>356.0</td>\n      <td>1.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>78.000000</td>\n      <td>78.000000</td>\n      <td>1017630</td>\n      <td>57.43850</td>\n      <td>10.55800</td>\n      <td>1697.0</td>\n      <td>-0.013064</td>\n      <td>11-47338586-C-T</td>\n      <td>0.0882</td>\n      <td>3.54</td>\n      <td>0.0</td>\n      <td>0.968143</td>\n      <td>1.537</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>402</th>\n      <td>402.0</td>\n      <td>0.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>74.863792</td>\n      <td>80.117728</td>\n      <td>1020263</td>\n      <td>383.80500</td>\n      <td>5.21918</td>\n      <td>1952.0</td>\n      <td>0.020625</td>\n      <td>11-47350068-C-T</td>\n      <td>0.1707</td>\n      <td>3.43</td>\n      <td>0.0</td>\n      <td>1.672304</td>\n      <td>0.511</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>572</th>\n      <td>572.0</td>\n      <td>0.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>67.370294</td>\n      <td>67.370294</td>\n      <td>1028911</td>\n      <td>-13.13840</td>\n      <td>1.26250</td>\n      <td>2791.0</td>\n      <td>-0.012814</td>\n      <td>11-47332912-A-G</td>\n      <td>0.3284</td>\n      <td>5.25</td>\n      <td>0.0</td>\n      <td>3.094806</td>\n      <td>7.144</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>726</th>\n      <td>726.0</td>\n      <td>0.0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>78.450376</td>\n      <td>78.450376</td>\n      <td>1036853</td>\n      <td>-5.87859</td>\n      <td>-4.38288</td>\n      <td>3548.0</td>\n      <td>-0.012519</td>\n      <td>11-47342016-G-A</td>\n      <td>0.6316</td>\n      <td>5.13</td>\n      <td>0.0</td>\n      <td>4.282963</td>\n      <td>2.437</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>502079</th>\n      <td>502079.0</td>\n      <td>0.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>77.864476</td>\n      <td>77.864476</td>\n      <td>6011707</td>\n      <td>-13.30810</td>\n      <td>2.01273</td>\n      <td>482436.0</td>\n      <td>0.015772</td>\n      <td>11-47342734-C-T</td>\n      <td>0.5758</td>\n      <td>4.63</td>\n      <td>0.0</td>\n      <td>3.708413</td>\n      <td>7.506</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>502084</th>\n      <td>502084.0</td>\n      <td>0.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>61.779603</td>\n      <td>61.779603</td>\n      <td>6011959</td>\n      <td>-11.68130</td>\n      <td>-2.81797</td>\n      <td>482460.0</td>\n      <td>-0.027074</td>\n      <td>11-47333208-C-T</td>\n      <td>0.4444</td>\n      <td>5.36</td>\n      <td>0.0</td>\n      <td>3.801036</td>\n      <td>7.556</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>502222</th>\n      <td>502222.0</td>\n      <td>1.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>61.535934</td>\n      <td>61.535934</td>\n      <td>6018974</td>\n      <td>-14.20980</td>\n      <td>2.78787</td>\n      <td>483140.0</td>\n      <td>-0.000138</td>\n      <td>11-47338579-G-A</td>\n      <td>0.1536</td>\n      <td>1.05</td>\n      <td>0.0</td>\n      <td>1.961707</td>\n      <td>0.428</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>502299</th>\n      <td>502299.0</td>\n      <td>0.0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>72.703628</td>\n      <td>72.703628</td>\n      <td>6022063</td>\n      <td>-12.70350</td>\n      <td>2.92162</td>\n      <td>483439.0</td>\n      <td>0.006107</td>\n      <td>11-47332912-A-G</td>\n      <td>0.3284</td>\n      <td>5.25</td>\n      <td>0.0</td>\n      <td>3.094806</td>\n      <td>7.144</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n    <tr>\n      <th>502342</th>\n      <td>502342.0</td>\n      <td>1.0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>77.032170</td>\n      <td>77.032170</td>\n      <td>6023765</td>\n      <td>-15.71110</td>\n      <td>1.60306</td>\n      <td>483600.0</td>\n      <td>-0.003158</td>\n      <td>11-47332912-A-G</td>\n      <td>0.3284</td>\n      <td>5.25</td>\n      <td>0.0</td>\n      <td>3.094806</td>\n      <td>7.144</td>\n      <td>0.0</td>\n      <td>1.0</td>\n    </tr>\n  </tbody>\n</table>\n<p>8044 rows \u00d7 19 columns</p>\n</div>"}, "metadata": {}}], "id": "8cc4eabb-6143-4c8e-9a18-10cc7e4f3b3b"}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": [], "id": "0a23b7d3-a747-4866-b75b-523541f571fa"}]}