{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.9.16"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# VEP API processing\n##### Updated 11/03/2023\n##### Selin Kubali\n\n#### Goal:\nThis notebook takes VCF files generated by vcfByGene and uses VEP API to add annotations predicting variants' conservation and deleteriousness.\n\n#### Required inputs\nVCF files with chrom, pos, ref, and alt information. Found in *selected_genes/hcm/vcf_files/*", "metadata": {"tags": []}, "id": "466b99bc-f7ca-4390-a4fe-8e1986414120"}, {"cell_type": "code", "source": "# download vcf files generated by vcfByGene\n!dx cd \"Cassa Lab Shared Project:selected_genes/hcm/vcf_files/\"\n!dx download *.vcf.gz", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": [{"name": "stdout", "output_type": "stream", "text": "^C2K[===>                                                        ] Downloaded 100,663,296 of 1,662,258,548 bytes (6%) /opt/notebooks/MYH6_variants.vcf.gz\n\n\n"}], "id": "d5cc5a9f-a2ac-41cb-942c-4dd5817d59eb"}, {"cell_type": "code", "source": "# unzip vcf files\n!gzip -d *.vcf.gz", "metadata": {"trusted": true, "tags": []}, "execution_count": 2, "outputs": [], "id": "5e79c70d-ddb9-4176-a21b-0ebbd3ba9d62"}, {"cell_type": "code", "source": "## cut out header and information related to patient genotypes from VCF files\n# you may have to switch to a bash kernel for this and switch back to a python kernel\n%%bash\nfor vcf_file in *.vcf; do\n  # Remove lines starting with '##' and then use 'cut' to extract fields 1-9\n  grep -v '^##' \"$vcf_file\" | cut -d$'\\t' -f 1-9 > \"${vcf_file%.vcf}_cleaned.vcf\"\ndone", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": [{"name": "stdout", "output_type": "stream", "text": "\n"}], "id": "70a9536d-7817-43d5-9376-ca1e419f7184"}, {"cell_type": "code", "source": "# upload unzipped and cut VCF files\n!dx cd \"Cassa Lab Shared Project:selected_genes/hcm/cleaned_vcf_files/\"\n!dx upload *_cleaned.vcf", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "e482dbf5-7b27-4057-96d0-4f11f42ced1f"}, {"cell_type": "code", "source": "# if the files are already uploaded in cleaned_vcf_files, skip the above steps and use this\n!dx cd \"Cassa Lab Shared Project:selected_genes/hcm/cleaned_vcf_files/\"\n!dx download *_variants_cleaned.vcf", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": [{"name": "stdout", "text": "\u001b[2K[===========================================================>] Completed 31,010 of 31,010 bytes (100%) /opt/notebooks/TTR_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 26,370 of 26,370 bytes (100%) /opt/notebooks/TNNI3_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 68,743 of 68,743 bytes (100%) /opt/notebooks/PTPN11_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 19,200 of 19,200 bytes (100%) /opt/notebooks/PLN_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 219,067 of 219,067 bytes (100%) /opt/notebooks/MYH7_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 248,239 of 248,239 bytes (100%) /opt/notebooks/MYH6_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 148,629 of 148,629 bytes (100%) /opt/notebooks/MYBPC3_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 234,222 of 234,222 bytes (100%) /opt/notebooks/FLNC_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 55,394 of 55,394 bytes (100%) /opt/notebooks/DES_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 199,347 of 199,347 bytes (100%) /opt/notebooks/ALPK3_variants_cleaned.vcff\n\u001b[2K[===========================================================>] Completed 139,897 of 139,897 bytes (100%) /opt/notebooks/ACTN2_variants_cleaned.vcff\n", "output_type": "stream"}], "id": "ddc420f1-1b84-4c9c-87ed-91f8b1aacd32"}, {"cell_type": "code", "source": "import pandas as pd\nimport numpy as np\nimport requests\nimport re\nimport sys\nfrom requests.exceptions import Timeout", "metadata": {"trusted": true, "tags": []}, "execution_count": 36, "outputs": [], "id": "76ab5fae-997d-4359-b726-306ee4a9a943"}, {"cell_type": "code", "source": "def find_max_val(gnomad_dict):\n    \"\"\" Loop through gnomad values and find highest gnomadg value \"\"\"\n    max_val = 0\n    for key in gnomad_dict:\n        if (gnomad_dict[key] > max_val) & (\"gnomadg\" in key):\n            max_val = gnomad_dict[key]\n    return max_val\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "2c7e4df9-c1a7-416c-a6bd-39a3ad46e6be"}, {"cell_type": "code", "source": "def find_max_af(decoded, alt):\n    \"\"\" Find gnomad values in JSON output \"\"\"\n    try:\n        gnomadg_s = decoded[0]['colocated_variants'][0]['frequencies'][alt]\n        gnomad_max_af = find_max_val(gnomadg_s)\n    except KeyError:\n        try:\n            gnomadg_s = decoded[0]['colocated_variants'][1]['frequencies'][alt]\n            gnomad_max_af = find_max_val(gnomadg_s)\n        except IndexError:\n            gnomad_max_af = None\n        except KeyError:\n            try:\n                gnomadg_s = decoded[0]['colocated_variants'][2]['frequencies'][alt]\n                gnomad_max_af = find_max_val(gnomadg_s)\n            except:\n                gnomad_max_af = None\n            \n    return gnomad_max_af\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 37, "outputs": [], "id": "00c05cac-a563-486f-8dba-49afd670b3e4"}, {"cell_type": "code", "source": "def find_trv(ref, alt):\n    \"\"\" Find whether transition or transversion \"\"\"\n    # transition = 0\n    # transversion = 1\n    trv = None\n    if (((ref == 'A') | (ref == 'G')) & ((alt == 'A') | (alt == 'G'))):\n        trv = 0\n    elif (((ref == 'C') | (ref == 'T')) & ((alt == 'C') | (alt == 'T'))):\n        trv = 0\n    else:\n        trv = 1\n    return trv", "metadata": {"trusted": true, "tags": []}, "execution_count": 38, "outputs": [{"name": "stdout", "text": "ERROR! Session/line number was not unique in database. History logging moved to new session 4\n", "output_type": "stream"}], "id": "95b387f2-764e-4237-bbe0-103947ffad92"}, {"cell_type": "code", "source": "def add_annotations(row, gene, df, df_row):\n    \n    \n    \"\"\" Takes a variant from .vcf files and returns information about variant from VEP API \"\"\"\n    # loads json with annotations for this variant\n    # note: works whether chr is int or chr# (eg 19 vs chr19, X vs chrX)\n\n    # parameters to request from API\n    parameters = {\n        \"species\": \"human\",\n        \"dbNSFP\": 'ALL',\n        \"assembly\": \"GRCh38\",\n        \"transcript_match\":1,\n        \"max_af\":1,\n        \"LoF\":1,\n        \"pick\":1,\n        \"pick_order\":\"canonical,appris,tsl,biotype,ccds,length,rank,mane_select,mane_plus_clinical\",\n        \"SpliceAI\":1\n    }\n\n\n    # send API request \n    try:\n        req = \"https://rest.ensembl.org/vep/human/hgvs/\" + f\"{row['#CHROM']}:g.{row['POS']}{row['REF']}>{row['ALT']}/?\" + \"&\".join([f\"{key}={value}\" for key, value in parameters.items()])\n        res = requests.get(req, headers={ \"Content-Type\" : \"application/json\"}, timeout=60)\n        \n        if not res.ok:\n          res.raise_for_status()\n          sys.exit()\n            \n        decoded = res.json()\n        \n\n    except requests.exceptions.HTTPError as e:\n \n        try:\n            len_ref = str(len(row['REF']) + row['POS'] - 1) # convert to as numeric\n            req = \"https://rest.ensembl.org/vep/human/hgvs/\" + f\"{row['#CHROM']}:g.{row['POS']}_{len_ref}del{row['REF']}ins{row['ALT']}/?\" + \"&\".join([f\"{key}={value}\" for key, value in parameters.items()])\n            res = requests.get(req, headers={ \"Content-Type\" : \"application/json\"}, timeout=60)\n\n            if not res.ok:\n              res.raise_for_status()\n              sys.exit()\n            decoded = res.json()\n        except Timeout:\n            print(\"Timeout error occurred. The request took too long to complete\")\n        \n        \n\n        \n\n    except Timeout:\n        try:\n            len_ref = str(len(row['REF']) + row['POS'] - 1) # convert to as numeric\n            req = \"https://rest.ensembl.org/vep/human/hgvs/\" + f\"{row['#CHROM']}:g.{row['POS']}_{len_ref}del{row['REF']}ins{row['ALT']}/?\" + \"&\".join([f\"{key}={value}\" for key, value in parameters.items()])\n            res = requests.get(req, headers={ \"Content-Type\" : \"application/json\"}, timeout=60)\n\n            if not res.ok:\n              res.raise_for_status()\n              sys.exit()\n            decoded = res.json()\n        except Timeout:\n            print(\"Timeout error occurred. The request took too long to complete\")\n        \n        \n\n   \n    ref = row['REF']\n    alt = row['ALT']\n    pos = row['POS']\n\n  \n    try:\n        ds_ag = decoded[0]['transcript_consequences'][0]['spliceai']['DS_AG']\n        ds_al = decoded[0]['transcript_consequences'][0]['spliceai']['DS_AL']\n        ds_dg = decoded[0]['transcript_consequences'][0]['spliceai']['DS_DG']\n        ds_dl = decoded[0]['transcript_consequences'][0]['spliceai']['DS_DL']\n    except KeyError:\n        ds_ag, ds_al, ds_dg, ds_dl = None, None, None, None\n\n\n\n    # extract relevant parameters\n    try:\n        phylop100way_vertebrate = decoded[0]['transcript_consequences'][0]['phylop100way_vertebrate']\n        gerp = decoded[0]['transcript_consequences'][0]['gerp++_rs']\n        cadd_raw = decoded[0]['transcript_consequences'][0]['cadd_raw']\n  \n    except KeyError:\n        phylop100way_vertebrate, gerp, cadd_raw  = None, None, None\n\n\n    trv = find_trv(ref, alt)\n    vep_consequence = decoded[0]['transcript_consequences'][0]['consequence_terms']\n    gene_id = decoded[0]['input']\n    chrom = re.search(r'^(.*?):', gene_id).group(1)\n    vep_consequence = \"&\".join(str(consequence) for consequence in vep_consequence)\n\n\n    # change formatting of ID\n    gene_id = gene_id.replace(\">\", \"-\" ) \n    gene_id = gene_id.replace(\":g.\", \"-\" ) \n    gene_id = gene_id[3:-3] + \"-\" + gene_id[-3:len(gene_id)]\n\n\n    # add parameters to df\n    df.loc[df_row, 'Name'] = gene_id\n    df.loc[df_row, 'Chrom'] = chrom\n    df.loc[df_row, 'Pos'] = pos\n    df.loc[df_row, 'Ref'] = ref\n    df.loc[df_row, 'Alt'] = alt\n    df.loc[df_row, 'vep_consequence'] = vep_consequence\n    df.loc[df_row, 'CADD_raw'] = cadd_raw\n    df.loc[df_row, 'phyloP100way_vertebrate'] = phylop100way_vertebrate\n    df.loc[df_row, 'GERP++_RS'] = gerp\n    df.loc[df_row, 'trv'] = trv\n    df.loc[df_row, 'ds_ag'] = ds_ag\n    df.loc[df_row, 'ds_al'] = ds_al\n    df.loc[df_row, 'ds_dg'] = ds_dg\n    df.loc[df_row, 'ds_dl'] = ds_dl\n    \n    \n    \n    \n    ", "metadata": {"trusted": true, "tags": []}, "execution_count": 39, "outputs": [], "id": "feb05173-d652-43b9-ac8c-0d434d01a39d"}, {"cell_type": "code", "source": "# loop through all rows in VCF; annotate with VEP; add to dataframe; and convert to CSV\n#genes =  [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\"]\ngenes =  [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\"]\n\nfor g in genes:\n    print(g)\n    parsed_df = pd.DataFrame({'Name': [], 'Chrom': [], 'Pos': [], 'Ref': [], 'Alt': [], 'vep_consequence': [], 'CADD_raw': [], 'phyloP100way_vertebrate': [], 'GERP++_RS': [], 'trv':[], 'ds_ag':[], 'ds_al':[], 'ds_dg': [], 'ds_dl': []})\n    gene_df = pd.read_csv(g + \"_variants_cleaned.vcf\", sep = \"\\t\")\n    \n    \n    for i in range(len(gene_df)):\n        add_annotations(gene_df.iloc[i,:],g,parsed_df, i)\n        print(i)\n        \n\n    parsed_df.to_csv(g+\".csv\")\n    \n", "metadata": {"trusted": true, "tags": []}, "execution_count": 42, "outputs": [{"name": "stdout", "text": "TTR\n0\n1\n2\n3\n4\n5\n6\n7\n8\n9\n10\n11\n12\n13\n14\n15\n16\n17\n18\n19\n20\n21\n22\n23\n24\n25\n26\n27\n28\n29\n30\n31\n32\n33\n34\n35\n36\n37\n38\n39\n40\n41\n42\n43\n44\n45\n46\n47\n48\n49\n50\n51\n52\n53\n54\n55\n56\n57\n58\n59\n60\n61\n62\n63\n64\n65\n66\n67\n68\n69\n70\n71\n72\n73\n74\n75\n76\n77\n78\n79\n80\n81\n82\n83\n84\n85\n86\n87\n88\n89\n90\n91\n92\n93\n94\n95\n96\n97\n98\n99\n100\n101\n102\n103\n104\n105\n106\n107\n108\n109\n110\n111\n112\n113\n114\n115\n116\n117\n118\n119\n120\n121\n122\n123\n124\n125\n126\n127\n128\n129\n130\n131\n132\n133\n134\n135\n136\n137\n138\n139\n140\n141\n142\n143\n144\n145\n146\n147\n148\n149\n150\n151\n152\n153\n154\n155\n156\n157\n158\n159\n160\n161\n162\n163\n164\n165\n166\n167\n168\n169\n170\n171\n172\n173\n174\n175\n", "output_type": "stream"}], "id": "26439e70-1970-47dc-8f3c-a17578bb2731"}, {"cell_type": "code", "source": "# upload processed VCFs\n!dx mkdir -p \"Cassa Lab Shared Project:selected_genes/hcm/parsed_vep_files/\"\n!dx cd /selected_genes/hcm/parsed_vep_files/\n!dx upload *.csv", "metadata": {"trusted": true, "tags": []}, "execution_count": 43, "outputs": [{"name": "stdout", "text": "[===========================================================>] Uploaded 76,258 of 76,258 bytes (100%) ACTN2.csv\nID                          file-GbggfZQJqBjB9J7739k3j1V9\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        ACTN2.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:42 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:43 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 106,325 of 106,325 bytes (100%) ALPK3.csv\nID                          file-Gbggfb0JqBjGq0yKypkvKgYV\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        ALPK3.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:44 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:45 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 29,609 of 29,609 bytes (100%) DES.csv\nID                          file-GbggfbQJqBj7zq7p7bX370Py\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        DES.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:46 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:47 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 126,345 of 126,345 bytes (100%) FLNC.csv\nID                          file-Gbggff0JqBj76JXb0v7YVgBK\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        FLNC.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:48 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:49 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 80,084 of 80,084 bytes (100%) MYBPC3.csv\nID                          file-GbggffQJqBj189P0F9vj7fV6\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        MYBPC3.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:50 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:52 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 133,084 of 133,084 bytes (100%) MYH6.csv\nID                          file-Gbggfg0JqBj52x2pzjVxX9ZK\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        MYH6.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:52 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:54 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 118,313 of 118,313 bytes (100%) MYH7.csv\nID                          file-GbggfgQJqBj91kpQ7G1XpXvf\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        MYH7.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:54 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:56 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 8,608 of 8,608 bytes (100%) PLN.csv\nID                          file-Gbggfj8JqBj9pk295ZF7vF8Q\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        PLN.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:57 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:34:58 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 37,570 of 37,570 bytes (100%) PTPN11.csv\nID                          file-GbggfjjJqBjPQq6ZVZYKGbvx\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        PTPN11.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:34:59 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:35:00 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 13,645 of 13,645 bytes (100%) TNNI3.csv\nID                          file-Gbggfk8JqBj0Yf1YfYpgk6B7\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        TNNI3.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:35:01 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:35:02 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n[===========================================================>] Uploaded 16,367 of 16,367 bytes (100%) TTR.csv\nID                          file-GbggfkjJqBjP7gfP2BKKqy7J\nClass                       file\nProject                     project-GGy3Bb0JqBj7zfxY8v4by61X\nFolder                      /selected_genes/hcm/parsed_vep_files\nName                        TTR.csv\nState                       \u001b[33mclosing\u001b[0m\nVisibility                  visible\nTypes                       -\nProperties                  -\nTags                        -\nOutgoing links              -\nCreated                     Tue Dec  5 19:35:03 2023\nCreated by                  skubali\n via the job                job-GbgX4XjJqBj3V7GyK8YGF1zj\nLast modified               Tue Dec  5 19:35:04 2023\nMedia type                  \narchivalState               \"live\"\ncloudAccount                \"cloudaccount-dnanexus\"\n", "output_type": "stream"}], "id": "eaf90121-8fca-4cdc-b123-f2fcf2ee494c"}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": [], "id": "3e4ab520-0474-4406-a50f-0f43de899abf"}]}